---
title: "UW Summer Institutes - Introduction to Metagenomic Data Analysis - Lab 6"
output: html_notebook
---

# Setup

```{r load_packages}
library(phyloseq)
library(vegan)
library(ade4)

library(ggplot2)
library(ggthemes)
library(stringr)
library(tidyr)
library(dplyr)
```

\  

# Load the provided RData image of Nature 488, pp. 621-626 data

```{r load_data}
load("../data/STAT.rdata")
```

\  

# Use `distance()` function to compute JSD distance of the normalized microbiome data.

```{r dist_jsd}
dist_jsd <- phyloseq::distance(phy, method = distanceMethodList$JSD)
str(dist_jsd)
```

```{r}
jsd.pco = dudi.pco(cailliez(dist_jsd), scannf=F, nf=2)

s.class(jsd.pco$li, sample_data(phy)$Location)
```

```{r}
s.class(jsd.pco$li, sample_data(phy)$Treatment)
```

```{r}
s.class(jsd.pco$li, with(sample_data(phy), 
                         interaction(Location,Treatment)))
```

# Use `adonis()` function of the `ade4` package to:

## Test for difference between the levels of Location variable.

```{r}
set.seed(0)
adonis(dist_jsd ~ Location, data = data.frame(sample_data(phy)), 
       permutations = 9999)
```


## Test for difference between the levels of Treatment variable.

```{r}
set.seed(0)
adonis(dist_jsd ~ Treatment, data = data.frame(sample_data(phy)), 
       permutations = 9999)
```

## Test for difference between the levels of Treatment variable stratified by Location.

```{r}
set.seed(0)
adonis(dist_jsd ~ Location+Treatment, data = data.frame(sample_data(phy)), 
       permutations = 9999)
```

```{r}
samp.data <- data.frame(sample_data(phy))
jsd.pco = dudi.pco(cailliez(dist_jsd), scannf=F, nf=2)
jsd.wca = wca(jsd.pco, samp.data$Location, scannf=F, nf=2)
s.class(jsd.wca$li, samp.data$Location)
```

```{r}
s.class(jsd.wca$li, samp.data$Treatment)
```

# Implement post hoc procedure for finding pairwise differences between Treatment levels in model 3.III. Compare these to post hoc tests without stratification.

```{r}
## Post-hoc tests
### Function for post-hoc test on two levels
## with stratification
Treatment = samp.data$Treatment
dist.mat = as.matrix(dist_jsd)
phoc.test = function(x){
    spair = (samp.data$Treatment %in% x)
    phoc = adonis(as.dist(dist.mat[spair,spair]) ~ Treatment[spair], 
                  data=samp.data, strata = samp.data$Location[spair])
    c(x, phoc$aov.tab[,6][1])
}
## Run the function above for all pairs of Treatment levels
combn(levels(Treatment),2, phoc.test )

### Function for post-hoc test on two levels
## without stratification
phoc.test2 = function(x){
    spair = (samp.data$Treatment %in% x)
    phoc = adonis(as.dist(dist.mat[spair,spair]) ~ Treatment[spair], 
                  data=samp.data)
    c(x, phoc$aov.tab[,6][1])
}
combn(levels(Treatment), 2, phoc.test2 )
```


# Within the subset of fecal samples compute the PERMANOVA analysis of Antibiotic (all types) vs. Control.

```{r}
sample_data(phy)$Abx = factor(Treatment != 'C')
fecal.phy = subset_samples(phy, Location == 'fecal')
fecal.dist = phyloseq::distance(fecal.phy, method="jsd")

adonis(fecal.dist~sample_data(fecal.phy)$Abx)
```

```{r}
sample_data(fecal.phy) %>% 
    data.frame() %>% 
    .[["Abx"]] %>% 
    table()
```


# Compare the significant values obtained in #5 to a simulated result obtained by running this code:

```{r}
simulate.ps = function(){
    n1 = 10; x1 = rnorm(n1, mean=0, sd = 10)
    n2 = 36; x2 = rnorm(n2, mean=0, sd = 1)
    x=c(x1,x2)
    f=factor(c(rep(1,n1), rep(2,n2)))
    dm = dist(x)
    c(adonis.p = adonis(dm~f)$aov.tab[1,6],
      tev.p = t.test(x1, x2, var.equal = T)$p.value,
      tuv.p = t.test(x1, x2, var.equal = F)$p.value)
}
set.seed(20160726)
simulate.ps()
sim = t(replicate(1000, simulate.ps()))
colSums(sim<0.05)/1000
```

# Given Tw2 statistic, implement permutation testing procedure and apply it to #5.

```{r}
WT = function(dm, f){
    if(nlevels(f) != 2) return(NULL)
    lev = levels(f)
    ns = table(f)
    N = sum(ns)
    dd = as.matrix(dm)
    dd[upper.tri(dd)]=0 ##
    SST = sum(dd^2)/N
    SSW1 = sum(dd[f==lev[1],f==lev[1]]^2)/ns[1]
    SSW2 = sum(dd[f==lev[2],f==lev[2]]^2)/ns[2]
    SSW = SSW1 + SSW2
    
    s1 = SSW1/(ns[1]-1)
    s2 = SSW2/(ns[2]-1)
    if(SST < SSW)
        t.stat = 0
    else
        t.stat =
        sqrt((ns[1]+ns[2])/(ns[1]*ns[2]))*sqrt(SST-SSW)/sqrt(s1/ns[1]
                                                            + s2/ns[2])
    t.stat
}
```

```{r}
WT.test = function(dm, f, nrep=999){
  stats = c(WT(dm, f), replicate(nrep, WT(dm, f[sample(length(f))])))
  p.value = sum(stats>=stats[1])/(nrep+1)
  t.stat = stats[1]
  list(p.value = p.value, t.stat = t.stat, nrep=nrep)  
}

WT.test(fecal.dist, sample_data(fecal.phy)$Abx)
```

```{r}
fecal.pco <- ordinate(fecal.phy, "PCoA", distance = fecal.dist)
plot_ordination(fecal.phy, fecal.pco, color = "Phylum", type = "byplot")
```


# Experiment with PERMANOVA-S, source code available at https://medschool.vanderbilt.edu/tang-lab/software/miProfile