---
title: "UW Summer Institutes - Introduction to Metagenomic Data Analysis - Lab 1"
output: html_notebook
---

# Getting ready

```{r load_packages, message=FALSE}
library(dada2); packageVersion("dada2")
library(ShortRead); packageVersion("ShortRead")
library(phyloseq); packageVersion("phyloseq")

library(ggplot2); packageVersion("ggplot2")
library(dplyr); packageVersion("dplyr")
```

```{r load_data}
path <- "../data/MiSeq_SOP/"
fns <- list.files(path)
fns
```

\  

# Filter, trim sequences

## Prepare sequence file names

```{r fnFs_fnRs}
fastqs <- fns[grepl("\\.fastq", fns)]
# Sort ensures forward/reverse files are in same order
fastqs <- sort(fastqs)
# Just the forward read files
fnFs <- fastqs[grepl("_R1", fastqs)]
# Just the reverse read files
fnRs <- fastqs[grepl("_R2", fastqs)]
# Get sample names from the first part of the forward read filenames
sample.names <- sapply(strsplit(fnFs, "_"), `[`, 1)
# Fully specify the path for the fnFs and fnRs
message("Forward read files:")
(fnFs <- paste0(path, fnFs))
message("Reverse read files:")
(fnRs <- paste0(path, fnRs))
```


## Examine quality profiles of forward and reverse reads

```{r plot_r1_quality_sample1}
plotQualityProfile(fnFs[[1]])
```

```{r plot_r2_quality_sample1}
plotQualityProfile(fnRs[[1]])
```

```{r plot_r1_quality_sample2}
plotQualityProfile(fnFs[[2]])
```

```{r plot_r2_quality_sample2}
plotQualityProfile(fnRs[[2]])
```

## Perform filtering and trimming

```{r filtFs_filtRs}
# Make filenames for the filtered fastq files
dirFiltSeqs = file.path(dirname(path), "filtered_sequences")
if(!dir.exists(dirFiltSeqs)){
  dir.create(dirFiltSeqs)
}
(filtFs <- file.path(dirFiltSeqs, paste0(sample.names, "_F_filt.fastq.gz")))
(filtRs <- file.path(dirFiltSeqs, paste0(sample.names, "_R_filt.fastq.gz")))
```

```{r filter_reads}
# Filter
for(i in seq_along(fnFs)) {
  fastqPairedFilter(fn = c(fnFs[i], fnRs[i]), 
                    fout = c(filtFs[i], filtRs[i]),
                    trimLeft = c(10, 10),
                    truncLen = c(240, 160), 
                    maxN = 0, 
                    maxEE = 2, 
                    truncQ = 2, 
                    compress = TRUE, verbose = TRUE)
}
```

**Q:** how does trimming maintain uniform read length?

\  

# Dereplicate sequences

**Note:** `derepFastq()` estimates a 'consensus quality' profile for each unique sequence by averaging positional qualities from dereplicated (duplicated) reads

```{r derepFs_derepRs}
derepFs <- derepFastq(filtFs, verbose=TRUE)
derepRs <- derepFastq(filtRs, verbose=TRUE)
# Name the derep-class objects by the sample names
names(derepFs) <- sample.names
names(derepRs) <- sample.names

derepFs[[1]]
derepRs[[1]]
```

\  

# Denoise sequences

```{r dadaFs}
dadaFs <- dada(derepFs, err=inflateErr(tperr1,3), selfConsist = TRUE)
```

```{r dadaRs}
dadaRs <- dada(derepRs, err=inflateErr(tperr1,3), selfConsist = TRUE)
```

```{r inspect_dadaFs}
dadaFs[[1]]
```

```{r plot_errors}
plotErrors(dadaFs[[1]], nominalQ=TRUE)
```

**Note:** estimated error rate is loess fit (I think?)

\  

# Merge paired reads

```{r mergers}
mergers <- mergePairs(dadaFs, derepFs, dadaRs, derepRs, verbose=TRUE)
# Inspect the merger data.frame from the first sample
head(mergers[[1]])
```

**Q:** with error removal by `dada()`, will `nindel` ever be > 0?

```{r}
sum(mergers[[1]]$nindel != 0)
```


\  

# Construct sequence table

```{r seqtab}
seqtab <- makeSequenceTable(mergers[names(mergers) != "Mock"])
dim(seqtab)
table(nchar(colnames(seqtab)))
```

**Q:** why does `derepFs` say that min, median, and max seq length is identical (230)?

\  

# Remove chimeras

```{r seqtab.nochim}
seqtab.nochim <- removeBimeraDenovo(seqtab, verbose=TRUE)
dim(seqtab.nochim)
sum(seqtab.nochim)/sum(seqtab)
```

\  

# Assign taxonomy

```{r taxa}
taxa <- assignTaxonomy(seqtab.nochim, "../data/MiSeq_SOP/rdp_train_set_14.fa.gz")
colnames(taxa) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus")
head(taxa)
unname(head(taxa))
```

\  

# Evaluate accuracy

```{r unqs.mock}
unqs.mock <- getUniques(removeBimeraDenovo(mergers[["Mock"]], verbose=TRUE))
cat("DADA2 inferred", length(unqs.mock), "sample sequences present in the Mock community.\n")
```

```{r match.ref}
mockRef <- readFasta("../data/MiSeq_SOP/HMP_MOCK.v35.fasta")
match.ref <- sum(sapply(names(unqs.mock), function(x) any(grepl(x, as.character(sread(mockRef))))))
cat("Of those,", sum(match.ref), "were exact matches to the expected reference sequences.\n")
```

\  

# Bonus: handoff to `phyloseq`

```{r samdf}
# Make a data.frame holding the sample data
samples.out <- rownames(seqtab.nochim)
subject <- sapply(strsplit(samples.out, "D"), `[`, 1)
gender <- substr(subject, 1, 1)
subject <- substr(subject, 2, 999)
day <- as.integer(sapply(strsplit(samples.out, "D"), `[`, 2))
samdf <- data.frame(Subject = subject, Gender = gender, Day = day)
samdf$When <- "Early"
samdf$When[samdf$Day > 100] <- "Late"
rownames(samdf) <- samples.out

glimpse(samdf)
```

## Import to `phyloseq`

```{r ps}
# Construct phyloseq object (straightforward from dada2 outputs)
ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows = FALSE), 
               sample_data(samdf), 
               tax_table(taxa))
ps
```

## Visualize alpha-diversity

```{r plot_richness}
plot_richness(ps, x="Day",
              measures=c("Shannon", "Simpson"), color="When") + 
    theme_bw()
```

## Ordinate

```{r ord.nmds.bray}
ord.nmds.bray <- ordinate(ps, method="NMDS", distance="bray")
```
```{r plot_ordination}
plot_ordination(ps, ord.nmds.bray, color="When", title="Bray NMDS")
```

```{r plot_bar}
top20 <- names(sort(taxa_sums(ps), decreasing=TRUE))[1:20]
ps.top20 <- transform_sample_counts(ps, function(OTU) OTU/sum(OTU))
ps.top20 <- prune_taxa(top20, ps.top20)
plot_bar(ps.top20, x="Day", fill="Family") + facet_wrap(~When, scales="free_x")
```


```{r plot_sample_sums}
plot(density(sample_sums(ps)))
```


