---
title: "Lab 03: Mixture Models for Testing Differential Abundance -- Example Solutions"
author: SISMID 2016 Module 14
date: "`r date()`"
output: 
  html_document: 
    fig_caption: yes
    number_sections: yes
    toc: yes
    toc_depth: 2
    toc_float: true
---


&nbsp;

Expected Time to Completion: 45 minutes

- [Simple Poisson simulation](Lab-03-Poisson-shot-noise.html)
- [DESeq2](http://bioconductor.org/packages/release/bioc/html/DESeq2.html)
- [edgeR](http://bioconductor.org/packages/release/bioc/html/edgeR.html)
- [metagenomeSeq](http://bioconductor.org/packages/release/bioc/html/metagenomeSeq.html)


&nbsp;

---

# Words of Caution

## The data here is a small toy 

The data here is what we call a "toy" example.
There are only relatively small number of taxa
compared to many common microbiome datasets, like human feces,
and these methods are not expected to behave optimally
in this current circumstance.

However, all the code "pieces" are in place,
so you can see how the "ins and outs" operate
and apply this to your own data.
Toy data give toy results!

## Check packages are installed

```{r}
packageVersion("phyloseq")
packageVersion("edgeR")
packageVersion("DESeq2")
packageVersion("metagenomeSeq")
```


If something is missing, use the Bioconductor installer:

```{r eval=FALSE}
## try http:// if https:// URLs are not supported
source("https://bioconductor.org/biocLite.R")
biocLite("edgeR")
```


---

&nbsp;

# The Poisson case

## "Shot noise" simulation

Time-permitting, I recommend that you review the included
[shot noise simulation tutorial](Lab-03-Poisson-shot-noise.html)
demonstrating the effect of sample size on species proportion estimates
in the simple Poisson case (no overdispersion).

This should be a quick read,
and a reminder why we need to include the *Library Size(s)*
in microbiome analyses, 
as the proportion value itself omits / masks
the implicit uncertainty of the sampling process.

The remainder of this lab is concerned 
with the more complicated case
in which we use a mixture model in order to account
for additional sources of uncertainty.


---

&nbsp;

# Startup R session

## Load phyloseq, ggplot2


```{r warning=FALSE, message=FALSE}
library("phyloseq"); packageVersion("phyloseq")
library("ggplot2"); packageVersion("ggplot2")
```

## Load provided example data

```{r}
load("example-data.RData")
lapply(lapply(ls(), get), nsamples)
lapply(lapply(ls(), get), ntaxa)
```

&nbsp;


---

# DESeq2

## DESeq2 with phyloseq

[DESeq2](http://www.biomedcentral.com/content/pdf/gb-2010-11-10-r106.pdf) 
has an official extension within the phyloseq package
and a vignette within phyloseq.

Anders S, Huber W (2010) Differential expression analysis for sequence count data.
[Genome Biology 11: R106](http://www.biomedcentral.com/content/pdf/gb-2010-11-10-r106.pdf)

## DESeq2 conversion

The function `phyloseq_to_deseq2` converts 
your phyloseq-format microbiome data into a `DESeqDataSet` 
with dispersions estimated, using the experimental design formula,
also shown (the `~Treatment` term). 
 
```{r run-deseq2}
dds = phyloseq_to_deseq2(closedps, ~ Treatment)
```

## DESeq2's DESeq function

The `DESeq` function does the rest of the modelling/testing. 
In this case we're using its default testing framework,
but you can use alternatives via `glm` interface.
The default multiple-inference correction is FDR,
and occurs within the `DESeq` function.

```{r deseq2, message=FALSE, warning=FALSE}
library("DESeq2"); packageVersion("DESeq2")
```
```{r}
dds = DESeq(dds, test="Wald", fitType="parametric")
```

## DESeq2 - explore results

`results` extracts a table of the test results from `dds`. Very fast. 
 The following will order by FDR, and remove `NA` values. 

```{r grab-results-process-table}
res = results(dds)
alpha = 0.05
# Add taxonomy to the results table
res1 = cbind(
  as(res, "data.frame"),
  as(tax_table(closedps)[rownames(res), ], "matrix"))
# Set NA `padj` values to max, 1.0
res1[is.na(res1$padj), "padj"] <- 1.0
sigtab = res1[which(res$padj < alpha), ]
```

Explore OTUs that were significantly different between the two classes.
The following makes a nice ggplot2 summary of the results.

To make the plot look nice,
we need to order things for ggplot2, 
and it typically does this by the order of elements in the factor.
This looks a bit complicated, but don't worry to much about it.
We are not changing the data at all here,
just the order of unique categories in each factor.

```{r plot-prep}
library("ggplot2")
theme_set(theme_bw())
plot_sigtab = function(dsdf){
  # Phylum order
  x = tapply(dsdf$log2FoldChange, dsdf$Phylum, function(x){max(x)})
  x = sort(x, TRUE)
  dsdf$Phylum <- factor(as.character(dsdf$Phylum), levels=names(x))
  # Genus order
  x = tapply(dsdf$log2FoldChange, dsdf$Genus, function(x){max(x)})
  x = sort(x, TRUE)
  dsdf$Genus <- factor(as.character(dsdf$Genus), levels=names(x))
  # Now define the ggplot2 object.
  p = ggplot(dsdf, 
             aes(x = log2FoldChange,
                 y = Genus,
                 color = Phylum,
                 size = -log10(padj),
                 shape = padj < 0.05)) + 
    geom_vline(xintercept = 0.0, linetype = 2) +
    geom_point(position = position_jitter(width = 0.1, height = 0.0)) + 
    scale_size_continuous(range = c(3, 7)) +
    theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5)) +
    ggtitle("DESeq2 Result")
  return(p)
}
pdsd = plot_sigtab(res1)
pdsd
```


## DESeq2 - Variance Stabilization

DESeq2 provides a variance stabilizing transformation (VST)
from the fitted dispersion-mean relation(s).
The count data is transformed and normalized by division by the library size factor,
yielding a matrix of values which are now approximately homoskedastic
(having constant variance along the range of mean values).
The `rlogTransformation` is less sensitive to size factors, 
which can be an issue when size factors vary widely. 
This transformation is useful when checking for outliers 
or as input for machine learning techniques such as clustering or linear discriminant analysis.

Relevant functions

- `varianceStabilizingTransformation`
- `getVarianceStabilizedData`
- `rlogTransformation` / `rlog`

Note that neither rlog transformation nor the VST 
are used by the differential expression estimation in DESeq,
which always occurs on the raw count data, 
through generalized linear modeling which incorporates knowledge of the variance-mean dependence. 
The rlog transformation and VST are offered as separate functionality 
which can be used for visualization, clustering or other machine learning tasks. 
See the transformation section of the vignette for more details.

## DESeq2 - VST

Perform VST, store the transformed values in a new table.

```{r vst}
vsd = getVarianceStabilizedData(dds)
dim(vsd)
cpsvsd = closedps
otu_table(cpsvsd) <- otu_table(vsd, taxa_are_rows = TRUE)
```

Compute and display ordination

```{r vst-plot}
pca = ordinate(cpsvsd, "RDA")
plot_ordination(cpsvsd, pca, type = "split", color="Phylum", shape="Treatment")
```


## DESeq2 - Regularized Log

We can do the same thing with regularized log.

The transformed values, `rlog(K)`, are equal to `rlog(K_ij) = log2(q_ij) = x_j. * beta_i`,
with formula terms defined in `DESeq` function documentation.

```{r warning=FALSE}
rld = rlog(dds, blind = FALSE)
cpsrld = closedps
otu_table(cpsrld) <- otu_table(assay(rld), taxa_are_rows = TRUE)
```



```{r rlog-plot}
pca = ordinate(cpsrld, "RDA")
plot_ordination(cpsrld, pca, type = "split",
                color="Phylum", shape="Treatment")
```


&nbsp;

---

# [edgeR](http://bioconductor.org/packages/release/bioc/html/edgeR.html)

## [edgeR Background](http://bioconductor.org/packages/release/bioc/html/edgeR.html)

Robinson MD, McCarthy DJ, Smyth GK (2009) edgeR: a Bioconductor package for differential expression analysis of digital gene expression data.
[Bioinformatics (Oxford, England) 26: 139â€“140](http://bioinformatics.oxfordjournals.org/content/26/1/139.short)

Another popular negative binomial method.

## Install edgeR

```{r install-edgeR, eval=FALSE}
source("http://bioconductor.org/biocLite.R")
biocLite("edgeR")
```


## Load package and function

```{r, message=FALSE, warning=FALSE}
library("edgeR"); packageVersion("edgeR")
```

Define a function for converting phyloseq data into edgeR DGE format.

```{r}
################################################################################
#' Convert phyloseq OTU count data into DGEList for edgeR package
#' 
#' Further details.
#' 
#' @param physeq (Required).  A \code{\link{phyloseq-class}} or
#'  an \code{\link{otu_table-class}} object. 
#'  The latter is only appropriate if \code{group} argument is also a 
#'  vector or factor with length equal to \code{nsamples(physeq)}.
#'  
#' @param group (Required). A character vector or factor giving the experimental
#'  group/condition for each sample/library. Alternatively, you may provide
#'  the name of a sample variable. This name should be among the output of
#'  \code{sample_variables(physeq)}, in which case
#'  \code{get_variable(physeq, group)} would return either a character vector or factor.
#'  This is passed on to \code{\link[edgeR]{DGEList}},
#'  and you may find further details or examples in its documentation.
#'  
#' @param method (Optional). The label of the edgeR-implemented normalization to use.
#'  See \code{\link[edgeR]{calcNormFactors}} for supported options and details. 
#'  The default option is \code{"RLE"}, which is a scaling factor method 
#'  proposed by Anders and Huber (2010).
#'  At time of writing, the \link[edgeR]{edgeR} package supported 
#'  the following options to the \code{method} argument:
#'  
#'  \code{c("TMM", "RLE", "upperquartile", "none")}.
#'
#' @param ... Additional arguments passed on to \code{\link[edgeR]{DGEList}}
#' 
#' @examples
#' 
phyloseq_to_edgeR = function(physeq, group, method="RLE", ...){
  require("edgeR")
  require("phyloseq")
  # Enforce orientation.
  if( !taxa_are_rows(physeq) ){ physeq <- t(physeq) }
  x = as(otu_table(physeq), "matrix")
  # Add one to protect against overflow, log(0) issues.
  x = x + 1
  # Check `group` argument
  if( identical(all.equal(length(group), 1), TRUE) & nsamples(physeq) > 1 ){
    # Assume that group was a sample variable name (must be categorical)
    group = get_variable(physeq, group)
  }
  # Define gene annotations (`genes`) as tax_table
  taxonomy = tax_table(physeq, errorIfNULL=FALSE)
  if( !is.null(taxonomy) ){
    taxonomy = data.frame(as(taxonomy, "matrix"))
  } 
  # Now turn into a DGEList
  y = DGEList(counts=x, group=group, genes=taxonomy, remove.zeros = TRUE, ...)
  # Calculate the normalization factors
  z = edgeR::calcNormFactors(y, method=method)
  # Check for division by zero inside `calcNormFactors`
  if( !all(is.finite(z$samples$norm.factors)) ){
    stop("Something wrong with edgeR::calcNormFactors on this data,
         non-finite $norm.factors, consider changing `method` argument")
  }
  # Estimate dispersions
  return(edgeR::estimateTagwiseDisp(edgeR::estimateCommonDisp(z)))
}
################################################################################
```


## edgeR - Quick Filtering

```{r}
cpsf = filter_taxa(closedps, function(x){sum(x>1) > (length(x)/4)}, prune = TRUE)
```


## edgeR - Convert for edgeR

Now let's use our newly-defined function to convert the phyloseq data object 
into an edgeR "DGE" data object, called `dge`.

```{r define-dge}
dge = phyloseq_to_edgeR(cpsf, group="Treatment")
```


## edgeR - Differential Abundance

Perform two-class test

```{r}
et = exactTest(dge)
```


## edgeR - Compile Results

Extract values and plot

```{r}
tt = topTags(et, n=nrow(dge$table), adjust.method="BH", sort.by="PValue")
res = tt@.Data[[1]]
resER = cbind(as(res, "data.frame"), 
              as(tax_table(closedps)[rownames(res), ], "matrix"))
sigtab = resER[(res$FDR < 0.05), ]
```


## edgeR - Plot

Differential Abundance

```{r, fig.width=11}
pedgeR = ggplot(resER,
                aes(logFC, Genus,
                    color = Phylum, 
                    size = -log10(FDR),
                    shape = FDR < 0.05)) + 
  geom_point() + 
  scale_size_continuous(range = c(3, 7)) +
  geom_vline(xintercept = 0) +
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0.5)) +
  ggtitle("edgeR Results")
print(pedgeR)
```


## [edgeR-robust](http://nar.oxfordjournals.org/content/42/11/e91.long)

Zhou, X., Lindsay, H., & Robinson, M. D. (2014). Robustly detecting differential expression in RNA sequencing data using observation weights. *Nucleic Acids Research*. 

`estimateGLMRobustDisp` - Empirical Robust Bayes Tagwise Dispersions for Negative Binomial GLMs using Observation Weights.

"At times, because of the moderation of dispersion estimates towards a trended values, features (typically, genes) can be sensitive to outliers and causing false positives. That is, since the dispersion estimates are moderated downwards toward the trend and because the regression parameter estimates may be affected by the outliers, genes are deemed significantly differential expressed"

```{r}
design <- model.matrix(~Treatment, data=data.frame(sample_data(cpsf))) # Define the design matrix for the full model
rer = estimateGLMRobustDisp(dge, design)
rertab <- topTags(exactTest(rer),
                  n=nrow(dge$table), 
                  adjust.method="BH",
                  sort.by="PValue")@.Data[[1]]
rertab = cbind(as(rertab, "data.frame"), 
      as(tax_table(cpsf)[rownames(rertab), ], "matrix"))
#rertab = rertab[(rertab$FDR < 0.05), ]
head(rertab)
```


## Plot (Robust)

We already defined the plot for edgeR earlier. 
Let's copy that plot, and then replace the data. Magic!

```{r, fig.width=11}
prer = pedgeR
prer$data <- rertab
prer <- prer + ggtitle("edgeR-robust")
print(prer)
```


## Detach edgeR package

We're going to move on to other packages. Let's "unload" edgeR
to avoid conflicts, namespace issues.

```{r}
unloadNamespace("edgeR")
```




&nbsp;

---

# metagenomeSeq

## metagenomeSeq - install

```{r eval=FALSE}
source("http://bioconductor.org/biocLite.R")
biocLite("metagenomeSeq")
```


## metagenomeSeq - Links and Background

[metagenomeSeq](http://cbcb.umd.edu/software/metagenomeSeq) is an official 
[Bioconductor package](http://www.bioconductor.org/packages/devel/bioc/html/metagenomeSeq.html), 
with the explicit goal of detecting differential abundance 
in microbiome experiments with an explicit design. 

First, we need functions to

- convert phyloseq data into a microbiomeSeq's `MGS` object
- perform the zero-inflated Gaussian fit 


## metagenomeSeq functions

Define a second function to 
based on our simple two-class experimental design.

Function to perform relevant test (`fitZig`)
- `MGS` - A metagenomeSeq object, most-likely produced using the conversion tool make_metagenomeSeq()
- `variable` - the variable among the sample_data (aka "phenoData" in metagenomeSeq) that you want to test
- `physeq` - optional phyloseq data object that has the relevant `tax_table`.
`phyloseq` or `taxonomyTable` is fine.

```{r metagenomeSeq-source}
library("metagenomeSeq"); packageVersion("metagenomeSeq")
# Function to convert form phyloseq object to metagenomeSeq object
phyloseq_to_metagenomeSeq = function(physeq){
  require("metagenomeSeq")
  require("phyloseq")
  # Enforce orientation
  if(!taxa_are_rows(physeq)){physeq <- t(physeq)}
  OTU = as(otu_table(physeq), "matrix")
  # Convert sample_data to AnnotatedDataFrame
  ADF = AnnotatedDataFrame(data.frame(sample_data(physeq)))
  # define dummy "feature" data for OTUs, using their name
  # Helps with extraction and relating to taxonomy later on.
  TDF = AnnotatedDataFrame(data.frame(OTUname = taxa_names(physeq), row.names=taxa_names(physeq)))
  # Create the metagenomeSeq object
  MGS = newMRexperiment(counts=OTU, phenoData=ADF, featureData=TDF)
  # Trigger metagenomeSeq to calculate its Cumulative Sum scaling factor.
  MGS = cumNorm(MGS)
  return(MGS)
}
# Function to perform relevant test (`fitZig`)
# - `MGS` - A metagenomeSeq object, most-likely produced using the conversion tool make_metagenomeSeq()
# - `variable` - the variable among the sample_data (aka "phenoData" in metagenomeSeq) that you want to test
# - `physeq` - optional phyloseq data object that has the relevant `tax_table`.
# `phyloseq` or `taxonomyTable` is fine.
test_metagenomeSeq = function(MGS, variable, physeq=NULL){
  require("metagenomeSeq")
  require("phyloseq")
  # Create the `mod` variable used in the fitZig test.
  if( inherits(variable, "factor") ){
    # If variable is already a factor, use directly in model.matrix
    mod = model.matrix( ~variable )
  } else if( inherits(variable, "matrix") ){
    # If it is a matrix, assume that model.matrix() has been used already
  } else if( inherits(variable, "character") ){ 
    # If it is a character that specifies a variable in phenoData, 
    # use the corresponding variable from MGS
    if( variable %in% colnames(phenoData(MGS)@data) ){
      mod = model.matrix(~phenoData(MGS)@data[, variable])
    } else {
      stop("The phenoData variable name you specified is not present in `phenoData(MGS)`")
    }
  } else {
    stop("Improper specification of the experimental design variable for testing. See `variable` argument")
  }
  # Wrapper to run the Expectation-maximization algorithm 
  # and estimate $f_count$ fits with the zero-inflated Guassian (z.i.g.)
  fit = fitZig(MGS, mod)
  # You need to specify all OTUs to get the full table from MRfulltable. 
  x = MRfulltable(fit, number=nrow(assayData(MGS)$counts))
  # if any OTUs left out, rm those from x. Detected by NA rownames.
  x = x[!is.na(rownames(x)), ]
  # Modify this data.frame by adding the OTUnames. Clip the ":1" added to the OTU names
  rownames(x) <- gsub(":1", "", x=rownames(x), fixed=TRUE)
  x$OTUnames <- as.character(rownames(x))
  if( !is.null(tax_table(physeq, errorIfNULL=FALSE)) ){
    # Attach the bacterial taxonomy to the table, if available
    TAX = data.frame(tax_table(physeq))
    TAX$OTUnames <- as.character(rownames(TAX))    
    y = merge(x, TAX, by="OTUnames") 
  } else {
    y = x
  }
  # Sort and return
  y = y[order(y$adjPvalue), ]
  return(y)
}
```


## metagenomeSeq - test

Perform metagenomeSeq differential abundance detection.
The default method for multiple-inference adjustment
in `metagenomeSeq::MRfulltable` is Benjamini-Hochberg (FDR).

```{r warning=FALSE}
mgs = phyloseq_to_metagenomeSeq(cpsf)
mgsres = test_metagenomeSeq(mgs, "Treatment", cpsf)
```


## metagenomeSeq - plot

```{r}
head(mgsres)
pmgs = ggplot(mgsres,
       aes(`(Intercept)`, Genus, color=adjPvalues < 0.05)) + 
  geom_point(position = position_jitter(width = 0.1, height = 0.0)) + 
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0.5)) + 
  ggtitle("metagenomeSeq")
print(pmgs)
```


## metagenomeSeq - Detach package

We're going to move on to other packages. Let's "unload" metagenomeSeq
to avoid conflicts, namespace issues.

```{r}
unloadNamespace("metagenomeSeq")
```



---

&nbsp;

# Compare Results

## Combining ggplot2 figures

The [gridExtra](http://cran.r-project.org/web/packages/gridExtra/index.html)
package adds some helpful utilities
for dealing with graphic objects created using
the low-level R graphics package called grid,
which is included as part of the base R distribution.
See [gridBase](http://cran.r-project.org/web/packages/gridBase/index.html)
for details about combining elements of grid graphics
with base R graphics.

## Combining ggplot2 figures

We will use the `grid.arrange` function from gridExtra

```{r}
library("gridExtra"); packageVersion("gridExtra")
```

## Combining ggplot2 figures BONUS!

```{r, fig.width=16, fig.height=16}
grid.arrange(nrow=2, pdsd, pedgeR, prer, pmgs)
```


