<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="SISMID 2016 Module 14" />


<title>Lab 03: Mixture Models for Testing Differential Abundance – Example Solutions</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<link rel="stylesheet"
      href="site_libs/highlight/default.css"
      type="text/css" />
<script src="site_libs/highlight/highlight.js"></script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>

<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.9em;
  padding-left: 5px;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Lab 03</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="http://sismid16m14.bot-tak.org/">
    <span class="fa fa-home"></span>
     
    home
  </a>
</li>
<li>
  <a href="Lab-03-mixture-models-Questions.html">
    <span class="fa fa-question"></span>
     
    Questions
  </a>
</li>
<li>
  <a href="Lab-03-mixture-models-Answers.html">
    <span class="fa fa-exclamation"></span>
     
    Answers
  </a>
</li>
<li>
  <a href="Lab-03-mixture-models-Answers.Rmd">
    <span class="fa fa-gear"></span>
     
    Source
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-newspaper-o"></span>
     
    Further Reading
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Lab-03-Poisson-shot-noise.html">Poisson Intuition Example</a>
    </li>
    <li>
      <a href="http://dx.doi.org/10.1371/journal.pcbi.1003531">Waste Not, Want Not</a>
    </li>
    <li class="dropdown-header">DESeq2</li>
    <li>
      <a href="http://www.biomedcentral.com/content/pdf/gb-2010-11-10-r106.pdf">DESeq2 article</a>
    </li>
    <li>
      <a href="http://bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2 package</a>
    </li>
    <li>
      <a href="http://www.biomedcentral.com/content/pdf/gb-2010-11-10-r106.pdf">DESeq original</a>
    </li>
    <li class="dropdown-header">edgeR</li>
    <li>
      <a href="http://bioinformatics.oxfordjournals.org/content/26/1/139.short">edgeR article</a>
    </li>
    <li>
      <a href="http://nar.oxfordjournals.org/content/early/2014/04/20/nar.gku310.full">edgeR-Robust</a>
    </li>
    <li>
      <a href="http://bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR package</a>
    </li>
    <li class="dropdown-header">metagenomeSeq</li>
    <li>
      <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4010126/">metagenomeSeq article</a>
    </li>
    <li>
      <a href="http://bioconductor.org/packages/release/bioc/html/metagenomeSeq.html">metagenomeSeq package</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Lab 03: Mixture Models for Testing Differential Abundance – Example Solutions</h1>
<h4 class="author"><em>SISMID 2016 Module 14</em></h4>
<h4 class="date"><em>Mon Jul 25 14:44:51 2016</em></h4>

</div>


<p> </p>
<p>Expected Time to Completion: 45 minutes</p>
<ul>
<li><a href="Lab-03-Poisson-shot-noise.html">Simple Poisson simulation</a></li>
<li><a href="http://bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2</a></li>
<li><a href="http://bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR</a></li>
<li><a href="http://bioconductor.org/packages/release/bioc/html/metagenomeSeq.html">metagenomeSeq</a></li>
</ul>
<p> </p>
<hr />
<div id="words-of-caution" class="section level1">
<h1><span class="header-section-number">1</span> Words of Caution</h1>
<div id="the-data-here-is-a-small-toy" class="section level2">
<h2><span class="header-section-number">1.1</span> The data here is a small toy</h2>
<p>The data here is what we call a “toy” example. There are only relatively small number of taxa compared to many common microbiome datasets, like human feces, and these methods are not expected to behave optimally in this current circumstance.</p>
<p>However, all the code “pieces” are in place, so you can see how the “ins and outs” operate and apply this to your own data. Toy data give toy results!</p>
</div>
<div id="check-packages-are-installed" class="section level2">
<h2><span class="header-section-number">1.2</span> Check packages are installed</h2>
<pre class="r"><code>packageVersion(&quot;phyloseq&quot;)</code></pre>
<pre><code>## [1] &#39;1.16.2&#39;</code></pre>
<pre class="r"><code>packageVersion(&quot;edgeR&quot;)</code></pre>
<pre><code>## [1] &#39;3.14.0&#39;</code></pre>
<pre class="r"><code>packageVersion(&quot;DESeq2&quot;)</code></pre>
<pre><code>## [1] &#39;1.12.3&#39;</code></pre>
<pre class="r"><code>packageVersion(&quot;metagenomeSeq&quot;)</code></pre>
<pre><code>## [1] &#39;1.14.2&#39;</code></pre>
<p>If something is missing, use the Bioconductor installer:</p>
<pre class="r"><code>## try http:// if https:// URLs are not supported
source(&quot;https://bioconductor.org/biocLite.R&quot;)
biocLite(&quot;edgeR&quot;)</code></pre>
<hr />
<p> </p>
</div>
</div>
<div id="the-poisson-case" class="section level1">
<h1><span class="header-section-number">2</span> The Poisson case</h1>
<div id="shot-noise-simulation" class="section level2">
<h2><span class="header-section-number">2.1</span> “Shot noise” simulation</h2>
<p>Time-permitting, I recommend that you review the included <a href="Lab-03-Poisson-shot-noise.html">shot noise simulation tutorial</a> demonstrating the effect of sample size on species proportion estimates in the simple Poisson case (no overdispersion).</p>
<p>This should be a quick read, and a reminder why we need to include the <em>Library Size(s)</em> in microbiome analyses, as the proportion value itself omits / masks the implicit uncertainty of the sampling process.</p>
<p>The remainder of this lab is concerned with the more complicated case in which we use a mixture model in order to account for additional sources of uncertainty.</p>
<hr />
<p> </p>
</div>
</div>
<div id="startup-r-session" class="section level1">
<h1><span class="header-section-number">3</span> Startup R session</h1>
<div id="load-phyloseq-ggplot2" class="section level2">
<h2><span class="header-section-number">3.1</span> Load phyloseq, ggplot2</h2>
<pre class="r"><code>library(&quot;phyloseq&quot;); packageVersion(&quot;phyloseq&quot;)</code></pre>
<pre><code>## [1] &#39;1.16.2&#39;</code></pre>
<pre class="r"><code>library(&quot;ggplot2&quot;); packageVersion(&quot;ggplot2&quot;)</code></pre>
<pre><code>## [1] &#39;2.1.0&#39;</code></pre>
</div>
<div id="load-provided-example-data" class="section level2">
<h2><span class="header-section-number">3.2</span> Load provided example data</h2>
<pre class="r"><code>load(&quot;example-data.RData&quot;)
lapply(lapply(ls(), get), nsamples)</code></pre>
<pre><code>## [[1]]
## [1] 9
## 
## [[2]]
## [1] 9
## 
## [[3]]
## [1] 26</code></pre>
<pre class="r"><code>lapply(lapply(ls(), get), ntaxa)</code></pre>
<pre><code>## [[1]]
## [1] 73
## 
## [[2]]
## [1] 419
## 
## [[3]]
## [1] 500</code></pre>
<p> </p>
<hr />
</div>
</div>
<div id="deseq2" class="section level1">
<h1><span class="header-section-number">4</span> DESeq2</h1>
<div id="deseq2-with-phyloseq" class="section level2">
<h2><span class="header-section-number">4.1</span> DESeq2 with phyloseq</h2>
<p><a href="http://www.biomedcentral.com/content/pdf/gb-2010-11-10-r106.pdf">DESeq2</a> has an official extension within the phyloseq package and a vignette within phyloseq.</p>
<p>Anders S, Huber W (2010) Differential expression analysis for sequence count data. <a href="http://www.biomedcentral.com/content/pdf/gb-2010-11-10-r106.pdf">Genome Biology 11: R106</a></p>
</div>
<div id="deseq2-conversion" class="section level2">
<h2><span class="header-section-number">4.2</span> DESeq2 conversion</h2>
<p>The function <code>phyloseq_to_deseq2</code> converts your phyloseq-format microbiome data into a <code>DESeqDataSet</code> with dispersions estimated, using the experimental design formula, also shown (the <code>~Treatment</code> term).</p>
<pre class="r"><code>dds = phyloseq_to_deseq2(closedps, ~ Treatment)</code></pre>
<pre><code>## Loading required namespace: DESeq2</code></pre>
<pre><code>## converting counts to integer mode</code></pre>
</div>
<div id="deseq2s-deseq-function" class="section level2">
<h2><span class="header-section-number">4.3</span> DESeq2’s DESeq function</h2>
<p>The <code>DESeq</code> function does the rest of the modelling/testing. In this case we’re using its default testing framework, but you can use alternatives via <code>glm</code> interface. The default multiple-inference correction is FDR, and occurs within the <code>DESeq</code> function.</p>
<pre class="r"><code>library(&quot;DESeq2&quot;); packageVersion(&quot;DESeq2&quot;)</code></pre>
<pre><code>## [1] &#39;1.12.3&#39;</code></pre>
<pre class="r"><code>dds = DESeq(dds, test=&quot;Wald&quot;, fitType=&quot;parametric&quot;)</code></pre>
<pre><code>## estimating size factors</code></pre>
<pre><code>## estimating dispersions</code></pre>
<pre><code>## gene-wise dispersion estimates</code></pre>
<pre><code>## mean-dispersion relationship</code></pre>
<pre><code>## final dispersion estimates</code></pre>
<pre><code>## fitting model and testing</code></pre>
</div>
<div id="deseq2---explore-results" class="section level2">
<h2><span class="header-section-number">4.4</span> DESeq2 - explore results</h2>
<p><code>results</code> extracts a table of the test results from <code>dds</code>. Very fast. The following will order by FDR, and remove <code>NA</code> values.</p>
<pre class="r"><code>res = results(dds)
alpha = 0.05
# Add taxonomy to the results table
res1 = cbind(
  as(res, &quot;data.frame&quot;),
  as(tax_table(closedps)[rownames(res), ], &quot;matrix&quot;))
# Set NA `padj` values to max, 1.0
res1[is.na(res1$padj), &quot;padj&quot;] &lt;- 1.0
sigtab = res1[which(res$padj &lt; alpha), ]</code></pre>
<p>Explore OTUs that were significantly different between the two classes. The following makes a nice ggplot2 summary of the results.</p>
<p>To make the plot look nice, we need to order things for ggplot2, and it typically does this by the order of elements in the factor. This looks a bit complicated, but don’t worry to much about it. We are not changing the data at all here, just the order of unique categories in each factor.</p>
<pre class="r"><code>library(&quot;ggplot2&quot;)
theme_set(theme_bw())
plot_sigtab = function(dsdf){
  # Phylum order
  x = tapply(dsdf$log2FoldChange, dsdf$Phylum, function(x){max(x)})
  x = sort(x, TRUE)
  dsdf$Phylum &lt;- factor(as.character(dsdf$Phylum), levels=names(x))
  # Genus order
  x = tapply(dsdf$log2FoldChange, dsdf$Genus, function(x){max(x)})
  x = sort(x, TRUE)
  dsdf$Genus &lt;- factor(as.character(dsdf$Genus), levels=names(x))
  # Now define the ggplot2 object.
  p = ggplot(dsdf, 
             aes(x = log2FoldChange,
                 y = Genus,
                 color = Phylum,
                 size = -log10(padj),
                 shape = padj &lt; 0.05)) + 
    geom_vline(xintercept = 0.0, linetype = 2) +
    geom_point(position = position_jitter(width = 0.1, height = 0.0)) + 
    scale_size_continuous(range = c(3, 7)) +
    theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5)) +
    ggtitle(&quot;DESeq2 Result&quot;)
  return(p)
}
pdsd = plot_sigtab(res1)
pdsd</code></pre>
<p><img src="Lab-03-mixture-models-Answers_files/figure-html/plot-prep-1.png" width="672" /></p>
</div>
<div id="deseq2---variance-stabilization" class="section level2">
<h2><span class="header-section-number">4.5</span> DESeq2 - Variance Stabilization</h2>
<p>DESeq2 provides a variance stabilizing transformation (VST) from the fitted dispersion-mean relation(s). The count data is transformed and normalized by division by the library size factor, yielding a matrix of values which are now approximately homoskedastic (having constant variance along the range of mean values). The <code>rlogTransformation</code> is less sensitive to size factors, which can be an issue when size factors vary widely. This transformation is useful when checking for outliers or as input for machine learning techniques such as clustering or linear discriminant analysis.</p>
<p>Relevant functions</p>
<ul>
<li><code>varianceStabilizingTransformation</code></li>
<li><code>getVarianceStabilizedData</code></li>
<li><code>rlogTransformation</code> / <code>rlog</code></li>
</ul>
<p>Note that neither rlog transformation nor the VST are used by the differential expression estimation in DESeq, which always occurs on the raw count data, through generalized linear modeling which incorporates knowledge of the variance-mean dependence. The rlog transformation and VST are offered as separate functionality which can be used for visualization, clustering or other machine learning tasks. See the transformation section of the vignette for more details.</p>
</div>
<div id="deseq2---vst" class="section level2">
<h2><span class="header-section-number">4.6</span> DESeq2 - VST</h2>
<p>Perform VST, store the transformed values in a new table.</p>
<pre class="r"><code>vsd = getVarianceStabilizedData(dds)
dim(vsd)</code></pre>
<pre><code>## [1] 73  9</code></pre>
<pre class="r"><code>cpsvsd = closedps
otu_table(cpsvsd) &lt;- otu_table(vsd, taxa_are_rows = TRUE)</code></pre>
<p>Compute and display ordination</p>
<pre class="r"><code>pca = ordinate(cpsvsd, &quot;RDA&quot;)
plot_ordination(cpsvsd, pca, type = &quot;split&quot;, color=&quot;Phylum&quot;, shape=&quot;Treatment&quot;)</code></pre>
<p><img src="Lab-03-mixture-models-Answers_files/figure-html/vst-plot-1.png" width="672" /></p>
</div>
<div id="deseq2---regularized-log" class="section level2">
<h2><span class="header-section-number">4.7</span> DESeq2 - Regularized Log</h2>
<p>We can do the same thing with regularized log.</p>
<p>The transformed values, <code>rlog(K)</code>, are equal to <code>rlog(K_ij) = log2(q_ij) = x_j. * beta_i</code>, with formula terms defined in <code>DESeq</code> function documentation.</p>
<pre class="r"><code>rld = rlog(dds, blind = FALSE)
cpsrld = closedps
otu_table(cpsrld) &lt;- otu_table(assay(rld), taxa_are_rows = TRUE)</code></pre>
<pre class="r"><code>pca = ordinate(cpsrld, &quot;RDA&quot;)
plot_ordination(cpsrld, pca, type = &quot;split&quot;,
                color=&quot;Phylum&quot;, shape=&quot;Treatment&quot;)</code></pre>
<p><img src="Lab-03-mixture-models-Answers_files/figure-html/rlog-plot-1.png" width="672" /></p>
<p> </p>
<hr />
</div>
</div>
<div id="edger" class="section level1">
<h1><span class="header-section-number">5</span> <a href="http://bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR</a></h1>
<div id="edger-background" class="section level2">
<h2><span class="header-section-number">5.1</span> <a href="http://bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR Background</a></h2>
<p>Robinson MD, McCarthy DJ, Smyth GK (2009) edgeR: a Bioconductor package for differential expression analysis of digital gene expression data. <a href="http://bioinformatics.oxfordjournals.org/content/26/1/139.short">Bioinformatics (Oxford, England) 26: 139–140</a></p>
<p>Another popular negative binomial method.</p>
</div>
<div id="install-edger" class="section level2">
<h2><span class="header-section-number">5.2</span> Install edgeR</h2>
<pre class="r"><code>source(&quot;http://bioconductor.org/biocLite.R&quot;)
biocLite(&quot;edgeR&quot;)</code></pre>
</div>
<div id="load-package-and-function" class="section level2">
<h2><span class="header-section-number">5.3</span> Load package and function</h2>
<pre class="r"><code>library(&quot;edgeR&quot;); packageVersion(&quot;edgeR&quot;)</code></pre>
<pre><code>## [1] &#39;3.14.0&#39;</code></pre>
<p>Define a function for converting phyloseq data into edgeR DGE format.</p>
<pre class="r"><code>################################################################################
#&#39; Convert phyloseq OTU count data into DGEList for edgeR package
#&#39; 
#&#39; Further details.
#&#39; 
#&#39; @param physeq (Required).  A \code{\link{phyloseq-class}} or
#&#39;  an \code{\link{otu_table-class}} object. 
#&#39;  The latter is only appropriate if \code{group} argument is also a 
#&#39;  vector or factor with length equal to \code{nsamples(physeq)}.
#&#39;  
#&#39; @param group (Required). A character vector or factor giving the experimental
#&#39;  group/condition for each sample/library. Alternatively, you may provide
#&#39;  the name of a sample variable. This name should be among the output of
#&#39;  \code{sample_variables(physeq)}, in which case
#&#39;  \code{get_variable(physeq, group)} would return either a character vector or factor.
#&#39;  This is passed on to \code{\link[edgeR]{DGEList}},
#&#39;  and you may find further details or examples in its documentation.
#&#39;  
#&#39; @param method (Optional). The label of the edgeR-implemented normalization to use.
#&#39;  See \code{\link[edgeR]{calcNormFactors}} for supported options and details. 
#&#39;  The default option is \code{&quot;RLE&quot;}, which is a scaling factor method 
#&#39;  proposed by Anders and Huber (2010).
#&#39;  At time of writing, the \link[edgeR]{edgeR} package supported 
#&#39;  the following options to the \code{method} argument:
#&#39;  
#&#39;  \code{c(&quot;TMM&quot;, &quot;RLE&quot;, &quot;upperquartile&quot;, &quot;none&quot;)}.
#&#39;
#&#39; @param ... Additional arguments passed on to \code{\link[edgeR]{DGEList}}
#&#39; 
#&#39; @examples
#&#39; 
phyloseq_to_edgeR = function(physeq, group, method=&quot;RLE&quot;, ...){
  require(&quot;edgeR&quot;)
  require(&quot;phyloseq&quot;)
  # Enforce orientation.
  if( !taxa_are_rows(physeq) ){ physeq &lt;- t(physeq) }
  x = as(otu_table(physeq), &quot;matrix&quot;)
  # Add one to protect against overflow, log(0) issues.
  x = x + 1
  # Check `group` argument
  if( identical(all.equal(length(group), 1), TRUE) &amp; nsamples(physeq) &gt; 1 ){
    # Assume that group was a sample variable name (must be categorical)
    group = get_variable(physeq, group)
  }
  # Define gene annotations (`genes`) as tax_table
  taxonomy = tax_table(physeq, errorIfNULL=FALSE)
  if( !is.null(taxonomy) ){
    taxonomy = data.frame(as(taxonomy, &quot;matrix&quot;))
  } 
  # Now turn into a DGEList
  y = DGEList(counts=x, group=group, genes=taxonomy, remove.zeros = TRUE, ...)
  # Calculate the normalization factors
  z = edgeR::calcNormFactors(y, method=method)
  # Check for division by zero inside `calcNormFactors`
  if( !all(is.finite(z$samples$norm.factors)) ){
    stop(&quot;Something wrong with edgeR::calcNormFactors on this data,
         non-finite $norm.factors, consider changing `method` argument&quot;)
  }
  # Estimate dispersions
  return(edgeR::estimateTagwiseDisp(edgeR::estimateCommonDisp(z)))
}
################################################################################</code></pre>
</div>
<div id="edger---quick-filtering" class="section level2">
<h2><span class="header-section-number">5.4</span> edgeR - Quick Filtering</h2>
<pre class="r"><code>cpsf = filter_taxa(closedps, function(x){sum(x&gt;1) &gt; (length(x)/4)}, prune = TRUE)</code></pre>
</div>
<div id="edger---convert-for-edger" class="section level2">
<h2><span class="header-section-number">5.5</span> edgeR - Convert for edgeR</h2>
<p>Now let’s use our newly-defined function to convert the phyloseq data object into an edgeR “DGE” data object, called <code>dge</code>.</p>
<pre class="r"><code>dge = phyloseq_to_edgeR(cpsf, group=&quot;Treatment&quot;)</code></pre>
</div>
<div id="edger---differential-abundance" class="section level2">
<h2><span class="header-section-number">5.6</span> edgeR - Differential Abundance</h2>
<p>Perform two-class test</p>
<pre class="r"><code>et = exactTest(dge)</code></pre>
</div>
<div id="edger---compile-results" class="section level2">
<h2><span class="header-section-number">5.7</span> edgeR - Compile Results</h2>
<p>Extract values and plot</p>
<pre class="r"><code>tt = topTags(et, n=nrow(dge$table), adjust.method=&quot;BH&quot;, sort.by=&quot;PValue&quot;)
res = tt@.Data[[1]]
resER = cbind(as(res, &quot;data.frame&quot;), 
              as(tax_table(closedps)[rownames(res), ], &quot;matrix&quot;))
sigtab = resER[(res$FDR &lt; 0.05), ]</code></pre>
</div>
<div id="edger---plot" class="section level2">
<h2><span class="header-section-number">5.8</span> edgeR - Plot</h2>
<p>Differential Abundance</p>
<pre class="r"><code>pedgeR = ggplot(resER,
                aes(logFC, Genus,
                    color = Phylum, 
                    size = -log10(FDR),
                    shape = FDR &lt; 0.05)) + 
  geom_point() + 
  scale_size_continuous(range = c(3, 7)) +
  geom_vline(xintercept = 0) +
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0.5)) +
  ggtitle(&quot;edgeR Results&quot;)
print(pedgeR)</code></pre>
<p><img src="Lab-03-mixture-models-Answers_files/figure-html/unnamed-chunk-12-1.png" width="1056" /></p>
</div>
<div id="edger-robust" class="section level2">
<h2><span class="header-section-number">5.9</span> <a href="http://nar.oxfordjournals.org/content/42/11/e91.long">edgeR-robust</a></h2>
<p>Zhou, X., Lindsay, H., &amp; Robinson, M. D. (2014). Robustly detecting differential expression in RNA sequencing data using observation weights. <em>Nucleic Acids Research</em>.</p>
<p><code>estimateGLMRobustDisp</code> - Empirical Robust Bayes Tagwise Dispersions for Negative Binomial GLMs using Observation Weights.</p>
<p>“At times, because of the moderation of dispersion estimates towards a trended values, features (typically, genes) can be sensitive to outliers and causing false positives. That is, since the dispersion estimates are moderated downwards toward the trend and because the regression parameter estimates may be affected by the outliers, genes are deemed significantly differential expressed”</p>
<pre class="r"><code>design &lt;- model.matrix(~Treatment, data=data.frame(sample_data(cpsf))) # Define the design matrix for the full model
rer = estimateGLMRobustDisp(dge, design)
rertab &lt;- topTags(exactTest(rer),
                  n=nrow(dge$table), 
                  adjust.method=&quot;BH&quot;,
                  sort.by=&quot;PValue&quot;)@.Data[[1]]
rertab = cbind(as(rertab, &quot;data.frame&quot;), 
      as(tax_table(cpsf)[rownames(rertab), ], &quot;matrix&quot;))
#rertab = rertab[(rertab$FDR &lt; 0.05), ]
head(rertab)</code></pre>
<pre><code>##          Kingdom          Phylum           Class             Order
## 270391  Bacteria   Bacteroidetes     Bacteroidia     Bacteroidales
## 4374042 Bacteria Deferribacteres Deferribacteres Deferribacterales
## 173807  Bacteria   Bacteroidetes     Bacteroidia     Bacteroidales
## 174126  Bacteria      Firmicutes      Clostridia     Clostridiales
## 259609  Bacteria   Bacteroidetes     Bacteroidia     Bacteroidales
## 182995  Bacteria   Bacteroidetes     Bacteroidia     Bacteroidales
##                       Family         Genus    Species     logFC   logCPM
## 270391         Rikenellaceae          &lt;NA&gt;       &lt;NA&gt;  2.860174 16.32399
## 4374042   Deferribacteraceae Mucispirillum schaedleri  2.385118 16.06727
## 173807  [Paraprevotellaceae]  [Prevotella]       &lt;NA&gt; -2.251202 16.28262
## 174126                  &lt;NA&gt;          &lt;NA&gt;       &lt;NA&gt; -2.642100 16.50780
## 259609                 S24-7          &lt;NA&gt;       &lt;NA&gt;  2.011105 15.72938
## 182995                 S24-7          &lt;NA&gt;       &lt;NA&gt;  1.447463 16.46517
##               PValue         FDR  Kingdom          Phylum           Class
## 270391  0.0002454804 0.004909609 Bacteria   Bacteroidetes     Bacteroidia
## 4374042 0.0019604404 0.019604404 Bacteria Deferribacteres Deferribacteres
## 173807  0.0049009896 0.032673264 Bacteria   Bacteroidetes     Bacteroidia
## 174126  0.0065561686 0.032780843 Bacteria      Firmicutes      Clostridia
## 259609  0.0124924325 0.049969730 Bacteria   Bacteroidetes     Bacteroidia
## 182995  0.0342079722 0.114026574 Bacteria   Bacteroidetes     Bacteroidia
##                     Order               Family         Genus    Species
## 270391      Bacteroidales        Rikenellaceae          &lt;NA&gt;       &lt;NA&gt;
## 4374042 Deferribacterales   Deferribacteraceae Mucispirillum schaedleri
## 173807      Bacteroidales [Paraprevotellaceae]  [Prevotella]       &lt;NA&gt;
## 174126      Clostridiales                 &lt;NA&gt;          &lt;NA&gt;       &lt;NA&gt;
## 259609      Bacteroidales                S24-7          &lt;NA&gt;       &lt;NA&gt;
## 182995      Bacteroidales                S24-7          &lt;NA&gt;       &lt;NA&gt;</code></pre>
</div>
<div id="plot-robust" class="section level2">
<h2><span class="header-section-number">5.10</span> Plot (Robust)</h2>
<p>We already defined the plot for edgeR earlier. Let’s copy that plot, and then replace the data. Magic!</p>
<pre class="r"><code>prer = pedgeR
prer$data &lt;- rertab
prer &lt;- prer + ggtitle(&quot;edgeR-robust&quot;)
print(prer)</code></pre>
<p><img src="Lab-03-mixture-models-Answers_files/figure-html/unnamed-chunk-14-1.png" width="1056" /></p>
</div>
<div id="detach-edger-package" class="section level2">
<h2><span class="header-section-number">5.11</span> Detach edgeR package</h2>
<p>We’re going to move on to other packages. Let’s “unload” edgeR to avoid conflicts, namespace issues.</p>
<pre class="r"><code>unloadNamespace(&quot;edgeR&quot;)</code></pre>
<p> </p>
<hr />
</div>
</div>
<div id="metagenomeseq" class="section level1">
<h1><span class="header-section-number">6</span> metagenomeSeq</h1>
<div id="metagenomeseq---install" class="section level2">
<h2><span class="header-section-number">6.1</span> metagenomeSeq - install</h2>
<pre class="r"><code>source(&quot;http://bioconductor.org/biocLite.R&quot;)
biocLite(&quot;metagenomeSeq&quot;)</code></pre>
</div>
<div id="metagenomeseq---links-and-background" class="section level2">
<h2><span class="header-section-number">6.2</span> metagenomeSeq - Links and Background</h2>
<p><a href="http://cbcb.umd.edu/software/metagenomeSeq">metagenomeSeq</a> is an official <a href="http://www.bioconductor.org/packages/devel/bioc/html/metagenomeSeq.html">Bioconductor package</a>, with the explicit goal of detecting differential abundance in microbiome experiments with an explicit design.</p>
<p>First, we need functions to</p>
<ul>
<li>convert phyloseq data into a microbiomeSeq’s <code>MGS</code> object</li>
<li>perform the zero-inflated Gaussian fit</li>
</ul>
</div>
<div id="metagenomeseq-functions" class="section level2">
<h2><span class="header-section-number">6.3</span> metagenomeSeq functions</h2>
<p>Define a second function to based on our simple two-class experimental design.</p>
<p>Function to perform relevant test (<code>fitZig</code>) - <code>MGS</code> - A metagenomeSeq object, most-likely produced using the conversion tool make_metagenomeSeq() - <code>variable</code> - the variable among the sample_data (aka “phenoData” in metagenomeSeq) that you want to test - <code>physeq</code> - optional phyloseq data object that has the relevant <code>tax_table</code>. <code>phyloseq</code> or <code>taxonomyTable</code> is fine.</p>
<pre class="r"><code>library(&quot;metagenomeSeq&quot;); packageVersion(&quot;metagenomeSeq&quot;)</code></pre>
<pre><code>## Loading required package: glmnet</code></pre>
<pre><code>## Loading required package: Matrix</code></pre>
<pre><code>## 
## Attaching package: &#39;Matrix&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:S4Vectors&#39;:
## 
##     expand</code></pre>
<pre><code>## Loading required package: foreach</code></pre>
<pre><code>## foreach: simple, scalable parallel programming from Revolution Analytics
## Use Revolution R for scalability, fault tolerance and more.
## http://www.revolutionanalytics.com</code></pre>
<pre><code>## Loaded glmnet 2.0-5</code></pre>
<pre><code>## Loading required package: RColorBrewer</code></pre>
<pre><code>## [1] &#39;1.14.2&#39;</code></pre>
<pre class="r"><code># Function to convert form phyloseq object to metagenomeSeq object
phyloseq_to_metagenomeSeq = function(physeq){
  require(&quot;metagenomeSeq&quot;)
  require(&quot;phyloseq&quot;)
  # Enforce orientation
  if(!taxa_are_rows(physeq)){physeq &lt;- t(physeq)}
  OTU = as(otu_table(physeq), &quot;matrix&quot;)
  # Convert sample_data to AnnotatedDataFrame
  ADF = AnnotatedDataFrame(data.frame(sample_data(physeq)))
  # define dummy &quot;feature&quot; data for OTUs, using their name
  # Helps with extraction and relating to taxonomy later on.
  TDF = AnnotatedDataFrame(data.frame(OTUname = taxa_names(physeq), row.names=taxa_names(physeq)))
  # Create the metagenomeSeq object
  MGS = newMRexperiment(counts=OTU, phenoData=ADF, featureData=TDF)
  # Trigger metagenomeSeq to calculate its Cumulative Sum scaling factor.
  MGS = cumNorm(MGS)
  return(MGS)
}
# Function to perform relevant test (`fitZig`)
# - `MGS` - A metagenomeSeq object, most-likely produced using the conversion tool make_metagenomeSeq()
# - `variable` - the variable among the sample_data (aka &quot;phenoData&quot; in metagenomeSeq) that you want to test
# - `physeq` - optional phyloseq data object that has the relevant `tax_table`.
# `phyloseq` or `taxonomyTable` is fine.
test_metagenomeSeq = function(MGS, variable, physeq=NULL){
  require(&quot;metagenomeSeq&quot;)
  require(&quot;phyloseq&quot;)
  # Create the `mod` variable used in the fitZig test.
  if( inherits(variable, &quot;factor&quot;) ){
    # If variable is already a factor, use directly in model.matrix
    mod = model.matrix( ~variable )
  } else if( inherits(variable, &quot;matrix&quot;) ){
    # If it is a matrix, assume that model.matrix() has been used already
  } else if( inherits(variable, &quot;character&quot;) ){ 
    # If it is a character that specifies a variable in phenoData, 
    # use the corresponding variable from MGS
    if( variable %in% colnames(phenoData(MGS)@data) ){
      mod = model.matrix(~phenoData(MGS)@data[, variable])
    } else {
      stop(&quot;The phenoData variable name you specified is not present in `phenoData(MGS)`&quot;)
    }
  } else {
    stop(&quot;Improper specification of the experimental design variable for testing. See `variable` argument&quot;)
  }
  # Wrapper to run the Expectation-maximization algorithm 
  # and estimate $f_count$ fits with the zero-inflated Guassian (z.i.g.)
  fit = fitZig(MGS, mod)
  # You need to specify all OTUs to get the full table from MRfulltable. 
  x = MRfulltable(fit, number=nrow(assayData(MGS)$counts))
  # if any OTUs left out, rm those from x. Detected by NA rownames.
  x = x[!is.na(rownames(x)), ]
  # Modify this data.frame by adding the OTUnames. Clip the &quot;:1&quot; added to the OTU names
  rownames(x) &lt;- gsub(&quot;:1&quot;, &quot;&quot;, x=rownames(x), fixed=TRUE)
  x$OTUnames &lt;- as.character(rownames(x))
  if( !is.null(tax_table(physeq, errorIfNULL=FALSE)) ){
    # Attach the bacterial taxonomy to the table, if available
    TAX = data.frame(tax_table(physeq))
    TAX$OTUnames &lt;- as.character(rownames(TAX))    
    y = merge(x, TAX, by=&quot;OTUnames&quot;) 
  } else {
    y = x
  }
  # Sort and return
  y = y[order(y$adjPvalue), ]
  return(y)
}</code></pre>
</div>
<div id="metagenomeseq---test" class="section level2">
<h2><span class="header-section-number">6.4</span> metagenomeSeq - test</h2>
<p>Perform metagenomeSeq differential abundance detection. The default method for multiple-inference adjustment in <code>metagenomeSeq::MRfulltable</code> is Benjamini-Hochberg (FDR).</p>
<pre class="r"><code>mgs = phyloseq_to_metagenomeSeq(cpsf)</code></pre>
<pre><code>## Default value being used.</code></pre>
<pre class="r"><code>mgsres = test_metagenomeSeq(mgs, &quot;Treatment&quot;, cpsf)</code></pre>
<pre><code>## it= 0, nll=13.81, log10(eps+1)=Inf, stillActive=20
## it= 1, nll=14.50, log10(eps+1)=0.06, stillActive=8
## it= 2, nll=14.46, log10(eps+1)=0.03, stillActive=5
## it= 3, nll=14.38, log10(eps+1)=0.05, stillActive=5
## it= 4, nll=14.41, log10(eps+1)=0.02, stillActive=4
## it= 5, nll=14.55, log10(eps+1)=0.00, stillActive=0</code></pre>
</div>
<div id="metagenomeseq---plot" class="section level2">
<h2><span class="header-section-number">6.5</span> metagenomeSeq - plot</h2>
<pre class="r"><code>head(mgsres)</code></pre>
<pre><code>##   OTUnames +samples in group 0 +samples in group 1 counts in group 0
## 1   169398                   4                   3                 9
## 2   173807                   5                   2                39
## 3   173923                   3                   3                 8
## 4   174126                   5                   0                31
## 5   182995                   5                   4                 8
## 6   183106                   4                   1                12
##   counts in group 1 oddsRatio       lower     upper     fisherP fisherAdjP
## 1                11 1.2909826 0.012781481 130.32575 1.000000000 1.00000000
## 2                 2       Inf 0.253802854       Inf 0.166666667 0.37037037
## 3                 5 0.5398122 0.006388843  16.49919 1.000000000 1.00000000
## 4                 0       Inf 1.750350751       Inf 0.007936508 0.05291005
## 5                21 0.0000000 0.000000000       Inf 1.000000000 1.00000000
## 6                 1 8.3550857 0.307130386 776.34824 0.206349206 0.41269841
##   (Intercept) phenoData(MGS)@data[, variable]Fast scalingFactor pvalues
## 1   1.0397871                           0.5101904      26.37353       1
## 2   2.2584200                          -1.6531739       9.93069       1
## 3  -0.5229753                          -0.1896061      95.73518       1
## 4   3.4376782                          -2.6181894     -43.89490       1
## 5   0.3489613                           1.2578970      48.99140       1
## 6   0.3499203                          -1.1795459      74.18522       1
##   adjPvalues  Kingdom        Phylum       Class         Order
## 1          1 Bacteria Bacteroidetes Bacteroidia Bacteroidales
## 2          1 Bacteria Bacteroidetes Bacteroidia Bacteroidales
## 3          1 Bacteria    Firmicutes  Clostridia Clostridiales
## 4          1 Bacteria    Firmicutes  Clostridia Clostridiales
## 5          1 Bacteria Bacteroidetes Bacteroidia Bacteroidales
## 6          1 Bacteria Bacteroidetes Bacteroidia Bacteroidales
##                 Family        Genus Species
## 1        Rikenellaceae         &lt;NA&gt;    &lt;NA&gt;
## 2 [Paraprevotellaceae] [Prevotella]    &lt;NA&gt;
## 3                 &lt;NA&gt;         &lt;NA&gt;    &lt;NA&gt;
## 4                 &lt;NA&gt;         &lt;NA&gt;    &lt;NA&gt;
## 5                S24-7         &lt;NA&gt;    &lt;NA&gt;
## 6                S24-7         &lt;NA&gt;    &lt;NA&gt;</code></pre>
<pre class="r"><code>pmgs = ggplot(mgsres,
       aes(`(Intercept)`, Genus, color=adjPvalues &lt; 0.05)) + 
  geom_point(position = position_jitter(width = 0.1, height = 0.0)) + 
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0.5)) + 
  ggtitle(&quot;metagenomeSeq&quot;)
print(pmgs)</code></pre>
<p><img src="Lab-03-mixture-models-Answers_files/figure-html/unnamed-chunk-18-1.png" width="672" /></p>
</div>
<div id="metagenomeseq---detach-package" class="section level2">
<h2><span class="header-section-number">6.6</span> metagenomeSeq - Detach package</h2>
<p>We’re going to move on to other packages. Let’s “unload” metagenomeSeq to avoid conflicts, namespace issues.</p>
<pre class="r"><code>unloadNamespace(&quot;metagenomeSeq&quot;)</code></pre>
<hr />
<p> </p>
</div>
</div>
<div id="compare-results" class="section level1">
<h1><span class="header-section-number">7</span> Compare Results</h1>
<div id="combining-ggplot2-figures" class="section level2">
<h2><span class="header-section-number">7.1</span> Combining ggplot2 figures</h2>
<p>The <a href="http://cran.r-project.org/web/packages/gridExtra/index.html">gridExtra</a> package adds some helpful utilities for dealing with graphic objects created using the low-level R graphics package called grid, which is included as part of the base R distribution. See <a href="http://cran.r-project.org/web/packages/gridBase/index.html">gridBase</a> for details about combining elements of grid graphics with base R graphics.</p>
</div>
<div id="combining-ggplot2-figures-1" class="section level2">
<h2><span class="header-section-number">7.2</span> Combining ggplot2 figures</h2>
<p>We will use the <code>grid.arrange</code> function from gridExtra</p>
<pre class="r"><code>library(&quot;gridExtra&quot;); packageVersion(&quot;gridExtra&quot;)</code></pre>
<pre><code>## 
## Attaching package: &#39;gridExtra&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:Biobase&#39;:
## 
##     combine</code></pre>
<pre><code>## The following object is masked from &#39;package:BiocGenerics&#39;:
## 
##     combine</code></pre>
<pre><code>## [1] &#39;2.2.1&#39;</code></pre>
</div>
<div id="combining-ggplot2-figures-bonus" class="section level2">
<h2><span class="header-section-number">7.3</span> Combining ggplot2 figures BONUS!</h2>
<pre class="r"><code>grid.arrange(nrow=2, pdsd, pedgeR, prer, pmgs)</code></pre>
<p><img src="Lab-03-mixture-models-Answers_files/figure-html/unnamed-chunk-21-1.png" width="1536" /></p>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
