---
title: "UW Summer Institutes - Introduction to Metagenomic Data Analysis - Lab 8"
output: html_notebook
---

# Setup

```{r load_packages}
library(phyloseq)

library(ggplot2)
library(ggthemes)
library(stringr)
library(tidyr)
library(dplyr)
```

\  

1. Load the predicted metagenome data from for Nature 488, pp. 621-626: STAT.picrust.RData.

```{r load_data}
load("../data/STAT.picrust.rdata")
```

\  


2. Study the provided data:
I. genes variable is a phyloseq object containing the metagenomic predictions (enzymes);

```{r inspect_genes}
str(genes)
str(sample_data(genes))
```

```{r plot_raw_counts}
otu_table(genes) %>% 
    as.data.frame() %>% 
    gather(sample, abundance) %>% 
    ggplot(aes(x = abundance)) +
    geom_density(aes(colour = sample)) +
    guides(colour = FALSE)
```


```{r plot_log_counts}
otu_table(genes) %>% 
    as.data.frame() %>% 
    gather(sample, abundance) %>% 
    ggplot(aes(x = log10(abundance + 1))) +
    geom_density(aes(colour = sample)) +
    guides(colour = FALSE)
```

```{r plot_rel_abundance}
transform_sample_counts(genes, function(x) x / sum(x)) %>% 
    otu_table() %>% 
    as.data.frame() %>% 
    gather(sample, abundance) %>% 
    ggplot(aes(x = abundance)) +
    geom_density(aes(colour = sample)) +
    guides(colour = FALSE)
```

```{r plot_filt_rel_abundance}
prune_taxa(taxa_sums(genes) > 0, genes) %>% 
    transform_sample_counts(function(x) x / sum(x)) %>% 
    otu_table() %>% 
    as.data.frame() %>% 
    gather(sample, abundance) %>% 
    ggplot(aes(x = abundance)) +
    geom_density(aes(colour = sample)) +
    guides(colour = FALSE)
```


II. pathways variable identifies biological pathways in which the metagenomic predictions participate. 

```{r inspect_pathways}
str(pathways)
head(pathways)
```

3. Use appropriate normalization technique to normalize the data.

```{r genes_norm}
genes_norm <- prune_taxa(taxa_sums(genes) > 0, genes) %>% 
    transform_sample_counts(function(x) x / sum(x))
str(genes_norm)
```

4. Perform principal components analysis ordination of the data, plot the results.

```{r genes_pcoa}
genes_pcoa <- ordinate(genes_norm, method = "PCoA", distance = "jsd")
```

```{r plot_pcoa_loc}
plot_ordination(genes_norm, genes_pcoa, color = "Location")
```

```{r plot_pcoa_tr}
plot_ordination(genes_norm, genes_pcoa, color = "Treatment")
```

5. Are these data multivariately associated with Location and/or Treatment? Which distance do you use for this analysis?

```{r jsd_pco}
library(vegan)
library(ade4)
genes_jsd <- phyloseq::distance(genes_norm, "jsd")
?distance

jsd_pco = dudi.pco(cailliez(genes_jsd), scannf=F, nf=2)
```

```{r plot_pco_loc}
s.class(jsd_pco$li, sample_data(genes_norm)$Location)
```

```{r plot_pco_tr}
s.class(jsd_pco$li, sample_data(genes_norm)$Treatment)
```


6. Perform univariate analysis of the metagenomic predictions with respect to the Location variable. What test do you use? Do not forget about adjustment for multiple comparisons! Report any positive results.

7. Fisher exact test, fisher.test, can be used to determine if a biological pathway is enriched for significant enzymes. Use this test on Level 1 pathways. Report any enriched pathways and perform the analysis at Level 2 for them. 

8. Perform the same analysis with respect to Treatment variable within the cecal samples. 
Note: use coefficient of variation cv = s.d./mean to filter out enzymes that do not vary much.  Keep only enzymes with cv>1.

What statistical test do you use to compare enzyme abundance in control and 
antibiotic groups?

Bonus: Load whole metagenome sequence data from Human Microbiome Project and 
experiment with these data (http://hmpdacc.org/HMMRC/).