<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="SISMID 2016 Module 14" />


<title>Lab 03: Shot Noise Poisson Simulation (for intuition)</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<link rel="stylesheet"
      href="site_libs/highlight/default.css"
      type="text/css" />
<script src="site_libs/highlight/highlight.js"></script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>

<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.9em;
  padding-left: 5px;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Lab 03</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="http://sismid16m14.bot-tak.org/">
    <span class="fa fa-home"></span>
     
    home
  </a>
</li>
<li>
  <a href="Lab-03-mixture-models-Questions.html">
    <span class="fa fa-question"></span>
     
    Questions
  </a>
</li>
<li>
  <a href="Lab-03-mixture-models-Answers.html">
    <span class="fa fa-exclamation"></span>
     
    Answers
  </a>
</li>
<li>
  <a href="Lab-03-mixture-models-Answers.Rmd">
    <span class="fa fa-gear"></span>
     
    Source
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-newspaper-o"></span>
     
    Further Reading
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Lab-03-Poisson-shot-noise.html">Poisson Intuition Example</a>
    </li>
    <li>
      <a href="http://dx.doi.org/10.1371/journal.pcbi.1003531">Waste Not, Want Not</a>
    </li>
    <li class="dropdown-header">DESeq2</li>
    <li>
      <a href="http://www.biomedcentral.com/content/pdf/gb-2010-11-10-r106.pdf">DESeq2 article</a>
    </li>
    <li>
      <a href="http://bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2 package</a>
    </li>
    <li>
      <a href="http://www.biomedcentral.com/content/pdf/gb-2010-11-10-r106.pdf">DESeq original</a>
    </li>
    <li class="dropdown-header">edgeR</li>
    <li>
      <a href="http://bioinformatics.oxfordjournals.org/content/26/1/139.short">edgeR article</a>
    </li>
    <li>
      <a href="http://nar.oxfordjournals.org/content/early/2014/04/20/nar.gku310.full">edgeR-Robust</a>
    </li>
    <li>
      <a href="http://bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR package</a>
    </li>
    <li class="dropdown-header">metagenomeSeq</li>
    <li>
      <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4010126/">metagenomeSeq article</a>
    </li>
    <li>
      <a href="http://bioconductor.org/packages/release/bioc/html/metagenomeSeq.html">metagenomeSeq package</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Lab 03: Shot Noise Poisson Simulation (for intuition)</h1>
<h4 class="author"><em>SISMID 2016 Module 14</em></h4>
<h4 class="date"><em>Mon Jul 25 14:45:14 2016</em></h4>

</div>


<p> </p>
<hr />
<div id="load-packages" class="section level1">
<h1><span class="header-section-number">1</span> Load packages</h1>
<pre class="r"><code>library(&quot;ggplot2&quot;); packageVersion(&quot;ggplot2&quot;)</code></pre>
<pre><code>## [1] &#39;2.1.0&#39;</code></pre>
<pre class="r"><code>theme_set(theme_bw())
library(&quot;plyr&quot;); packageVersion(&quot;plyr&quot;)</code></pre>
<pre><code>## 
## Attaching package: &#39;plyr&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:IRanges&#39;:
## 
##     desc</code></pre>
<pre><code>## The following object is masked from &#39;package:S4Vectors&#39;:
## 
##     rename</code></pre>
<pre><code>## [1] &#39;1.8.4&#39;</code></pre>
<pre class="r"><code>library(&quot;data.table&quot;); packageVersion(&quot;data.table&quot;)</code></pre>
<pre><code>## data.table 1.9.6  For help type ?data.table or https://github.com/Rdatatable/data.table/wiki</code></pre>
<pre><code>## The fastest way to learn (by data.table authors): https://www.datacamp.com/courses/data-analysis-the-data-table-way</code></pre>
<pre><code>## 
## Attaching package: &#39;data.table&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:SummarizedExperiment&#39;:
## 
##     shift</code></pre>
<pre><code>## The following object is masked from &#39;package:GenomicRanges&#39;:
## 
##     shift</code></pre>
<pre><code>## The following object is masked from &#39;package:IRanges&#39;:
## 
##     shift</code></pre>
<pre><code>## [1] &#39;1.9.6&#39;</code></pre>
<pre class="r"><code>set.seed(711L)</code></pre>
<p> </p>
<hr />
</div>
<div id="define-simple-microbiome" class="section level1">
<h1><span class="header-section-number">2</span> Define simple microbiome</h1>
<p>Let’s define an imaginary and simple vector of proportions of bacterial taxa in a community.</p>
<pre class="r"><code>OTUproportions = c(0.3, 0.2, 0.1,
                   rep(0.05, 5), 
                   rep(0.025, 6), 
                   rep(0.001, 6), 
                   rep(0.0001, 6))
names(OTUproportions) &lt;- LETTERS[1:length(OTUproportions)]
OTUproportions</code></pre>
<pre><code>##      A      B      C      D      E      F      G      H      I      J 
## 0.3000 0.2000 0.1000 0.0500 0.0500 0.0500 0.0500 0.0500 0.0250 0.0250 
##      K      L      M      N      O      P      Q      R      S      T 
## 0.0250 0.0250 0.0250 0.0250 0.0010 0.0010 0.0010 0.0010 0.0010 0.0010 
##      U      V      W      X      Y      Z 
## 0.0001 0.0001 0.0001 0.0001 0.0001 0.0001</code></pre>
<pre class="r"><code>sum(OTUproportions)</code></pre>
<pre><code>## [1] 1.0066</code></pre>
<pre class="r"><code># Plot it to be clear.
spdf = data.frame(proportion=OTUproportions, Taxa=names(OTUproportions))
p = ggplot(spdf, aes(Taxa, proportion)) + geom_bar(stat=&quot;identity&quot;) 
p + ggtitle(&quot;The &#39;true&#39; Taxa proportions&quot;)</code></pre>
<p><img src="Lab-03-Poisson-shot-noise_files/figure-html/unnamed-chunk-1-1.png" width="672" /></p>
<p>Now sample from these true proportions, and plot the uncertainty. This uses the <code>sample</code> function in R.</p>
<pre class="r"><code>sim1 = sample(x=names(OTUproportions), size=100, replace=TRUE, prob=OTUproportions)
sim1</code></pre>
<pre><code>##   [1] &quot;A&quot; &quot;B&quot; &quot;B&quot; &quot;H&quot; &quot;B&quot; &quot;A&quot; &quot;B&quot; &quot;A&quot; &quot;C&quot; &quot;E&quot; &quot;A&quot; &quot;H&quot; &quot;L&quot; &quot;D&quot; &quot;I&quot; &quot;I&quot; &quot;F&quot;
##  [18] &quot;F&quot; &quot;M&quot; &quot;G&quot; &quot;A&quot; &quot;B&quot; &quot;B&quot; &quot;A&quot; &quot;A&quot; &quot;A&quot; &quot;H&quot; &quot;H&quot; &quot;A&quot; &quot;D&quot; &quot;C&quot; &quot;D&quot; &quot;A&quot; &quot;F&quot;
##  [35] &quot;I&quot; &quot;E&quot; &quot;B&quot; &quot;C&quot; &quot;D&quot; &quot;E&quot; &quot;A&quot; &quot;A&quot; &quot;B&quot; &quot;B&quot; &quot;M&quot; &quot;E&quot; &quot;C&quot; &quot;A&quot; &quot;A&quot; &quot;I&quot; &quot;D&quot;
##  [52] &quot;E&quot; &quot;M&quot; &quot;C&quot; &quot;F&quot; &quot;B&quot; &quot;G&quot; &quot;E&quot; &quot;H&quot; &quot;A&quot; &quot;B&quot; &quot;B&quot; &quot;A&quot; &quot;A&quot; &quot;F&quot; &quot;B&quot; &quot;A&quot; &quot;N&quot;
##  [69] &quot;G&quot; &quot;F&quot; &quot;A&quot; &quot;B&quot; &quot;A&quot; &quot;H&quot; &quot;A&quot; &quot;A&quot; &quot;A&quot; &quot;H&quot; &quot;H&quot; &quot;H&quot; &quot;D&quot; &quot;A&quot; &quot;H&quot; &quot;F&quot; &quot;B&quot;
##  [86] &quot;A&quot; &quot;K&quot; &quot;G&quot; &quot;A&quot; &quot;A&quot; &quot;A&quot; &quot;G&quot; &quot;C&quot; &quot;B&quot; &quot;A&quot; &quot;B&quot; &quot;G&quot; &quot;Q&quot; &quot;B&quot; &quot;B&quot;</code></pre>
<pre class="r"><code>sim1df = data.frame(table(sim1))
sim1df</code></pre>
<pre><code>##    sim1 Freq
## 1     A   29
## 2     B   19
## 3     C    6
## 4     D    6
## 5     E    6
## 6     F    7
## 7     G    6
## 8     H   10
## 9     I    4
## 10    K    1
## 11    L    1
## 12    M    3
## 13    N    1
## 14    Q    1</code></pre>
<pre class="r"><code>sim1df$proportion &lt;- sim1df$Freq/sum(sim1df$Freq)
colnames(sim1df)[1] &lt;- &quot;Taxa&quot;
# Add simulated proportions to the plot of the true proportions.
title = &quot;One realization of multinomial sampling process&quot;
p + geom_point(data=sim1df, color=&quot;blue&quot;, size=6) +
  ggtitle(title)</code></pre>
<p><img src="Lab-03-Poisson-shot-noise_files/figure-html/simulate-1.png" width="672" /></p>
<p> </p>
<hr />
</div>
<div id="define-simulation" class="section level1">
<h1><span class="header-section-number">3</span> Define simulation</h1>
<p>Define and test a function that simulates from this</p>
<pre class="r"><code>simulate_community = function(props, size){
  # props = OTUproportions
  # size = 100
  simi = sample(x=names(props), size=size, replace=TRUE, prob=props)
  simidf = data.table(table(simi))
  simidf[, proportion := N / sum(N, na.rm = TRUE)]
  setnames(simidf, &quot;simi&quot;, &quot;Taxa&quot;)
  return(simidf)
}
# Test the function
show(simulate_community(OTUproportions, 200))</code></pre>
<pre><code>##     Taxa  N proportion
##  1:    A 69      0.345
##  2:    B 35      0.175
##  3:    C 15      0.075
##  4:    D  7      0.035
##  5:    E  8      0.040
##  6:    F  9      0.045
##  7:    G 10      0.050
##  8:    H 16      0.080
##  9:    I  7      0.035
## 10:    J  6      0.030
## 11:    K  4      0.020
## 12:    L  4      0.020
## 13:    M  6      0.030
## 14:    N  4      0.020</code></pre>
<pre class="r"><code># Repeat twice, store as list
replicate(2, simulate_community(OTUproportions, 100), simplify = FALSE)</code></pre>
<pre><code>## [[1]]
##     Taxa  N proportion
##  1:    A 29       0.29
##  2:    B 23       0.23
##  3:    C  9       0.09
##  4:    D  4       0.04
##  5:    E  4       0.04
##  6:    F  5       0.05
##  7:    G  3       0.03
##  8:    H  5       0.05
##  9:    J  2       0.02
## 10:    K  4       0.04
## 11:    L  5       0.05
## 12:    M  4       0.04
## 13:    N  2       0.02
## 14:    Q  1       0.01
## 
## [[2]]
##     Taxa  N proportion
##  1:    A 27       0.27
##  2:    B 21       0.21
##  3:    C  7       0.07
##  4:    D  2       0.02
##  5:    E  9       0.09
##  6:    F  2       0.02
##  7:    G  6       0.06
##  8:    H  8       0.08
##  9:    I  2       0.02
## 10:    J  5       0.05
## 11:    K  2       0.02
## 12:    L  1       0.01
## 13:    M  5       0.05
## 14:    N  3       0.03</code></pre>
<p> </p>
<hr />
</div>
<div id="repeat-simulation" class="section level1">
<h1><span class="header-section-number">4</span> Repeat Simulation</h1>
<p>Now repeat the simulated Taxa sampling many times at different <em>library sizes</em>.</p>
<pre class="r"><code>N = 300
LibrarySizes = c(100, 2000, 30000, 400000)
bigdt = rbindlist(lapply(X = LibrarySizes, FUN = function(libsz){
  listi = replicate(N, simulate_community(OTUproportions, libsz), simplify=FALSE)
  dti = rbindlist(listi)
  dti[, LibrarySize := as.character(libsz)]
  return(dti)
}))</code></pre>
<p> </p>
<hr />
</div>
<div id="graphic-summaries" class="section level1">
<h1><span class="header-section-number">5</span> Graphic Summaries</h1>
<div id="scatter-plot-of-individual-results" class="section level2">
<h2><span class="header-section-number">5.1</span> Scatter plot of individual results</h2>
<pre class="r"><code>title = &quot;Accuracy of Individual Simulations&quot;
p + geom_point(aes(color=LibrarySize, fill=LibrarySize), data=bigdt,
               position=position_jitterdodge(), size=1) + 
  ggtitle(title)</code></pre>
<p><img src="Lab-03-Poisson-shot-noise_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
<pre class="r"><code>title = &quot;Accuracy of Aggregate Simulations (boxplot)&quot;
p + geom_boxplot(aes(fill=LibrarySize),
                 data=bigdt, 
                 position=&quot;dodge&quot;) +
  ggtitle(title)</code></pre>
<p><img src="Lab-03-Poisson-shot-noise_files/figure-html/unnamed-chunk-4-2.png" width="672" /></p>
<p> </p>
</div>
<div id="residuals-v.-counts" class="section level2">
<h2><span class="header-section-number">5.2</span> Residuals v. Counts</h2>
<pre class="r"><code>bigdt[, TrueProportion := spdf[Taxa, &quot;proportion&quot;]]
bigdt[, Residual := proportion - TrueProportion]
bigdt[, Error := abs(Residual)]
bigdt[, StdErr := Residual / proportion]</code></pre>
<pre class="r"><code>ggplot(bigdt, aes(x = N,
                  y = Residual,
                  color = factor(TrueProportion),
                  shape = LibrarySize)) + 
  geom_point(size = 2, alpha = 0.5) + 
  scale_x_log10() + 
  annotation_logticks(sides = &quot;b&quot;) +
  ggtitle(&quot;Residual v. Absolute Count&quot;)</code></pre>
<p><img src="Lab-03-Poisson-shot-noise_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<p> </p>
</div>
<div id="variance-v.-counts" class="section level2">
<h2><span class="header-section-number">5.3</span> Variance v. Counts</h2>
<p>Graphically summarize variance, standard deviation, and error.</p>
<p>First prepare data for plotting</p>
<pre class="r"><code># Summarize by trial and Taxa
bigdt2 = ddply(bigdt, c(&quot;LibrarySize&quot;, &quot;Taxa&quot;), function(x){
  data.frame(Variance = var(x$N),
             standardDeviation = sd(x$N),
             meanCount = mean(x$N),
             meanProp = mean(x$proportion),
             L = as.integer(x$LibrarySize[1])
             )
})
bigdt2 &lt;- data.table(bigdt2)
bigdt2[, TrueProportion := spdf[Taxa, &quot;proportion&quot;]]
bigdt2[, cv := Variance / as.numeric(LibrarySize)]</code></pre>
<p>Define plots</p>
<pre class="r"><code># Plots
title = &quot;Accuracy and Library Size&quot;
p + geom_boxplot(aes(fill=LibrarySize), data=bigdt, position=&quot;dodge&quot;) +
  geom_errorbar(data=bigdt2, position=&quot;dodge&quot;,
                aes(x=Taxa, y=NULL, colour=LibrarySize,
                    ymax=meanProp+standardDeviation/L,
                    ymin=meanProp-standardDeviation/L)) + 
  ggtitle(title)</code></pre>
<pre><code>## Warning: Removed 2 rows containing missing values (geom_errorbar).</code></pre>
<p><img src="Lab-03-Poisson-shot-noise_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
<pre class="r"><code>title = &quot;Observed Variance versus Observed Mean Counts&quot;
ggplot(data = bigdt2, 
       mapping = aes(x = meanCount,
                     y = Variance,
                     colour=LibrarySize)) + 
  geom_point(size=5, alpha=0.5) + 
  geom_abline(intercept = 0, slope = 1) + 
  scale_y_log10() + 
  scale_x_log10() + 
  annotation_logticks(sides = &quot;lb&quot;) +
  ggtitle(title)</code></pre>
<pre><code>## Warning: Removed 2 rows containing missing values (geom_point).</code></pre>
<p><img src="Lab-03-Poisson-shot-noise_files/figure-html/unnamed-chunk-8-2.png" width="672" /></p>
<pre class="r"><code>title = &quot;Observed Variance versus Expected Variance (Poisson)&quot;
ggplot(bigdt2, aes(L*TrueProportion, Variance, color=LibrarySize)) + geom_point(size=5, alpha=0.5) + 
  geom_abline(intercept = 0, slope = 1) + 
  scale_y_log10() + 
  scale_x_log10() + 
  annotation_logticks(sides = &quot;lb&quot;) +
  xlab(&quot;Expected Variance [counts^2]&quot;) + 
  ylab(&quot;Estimated Variance [counts^2]&quot;) +
  ggtitle(title)</code></pre>
<pre><code>## Warning: Removed 2 rows containing missing values (geom_point).</code></pre>
<p><img src="Lab-03-Poisson-shot-noise_files/figure-html/unnamed-chunk-8-3.png" width="672" /></p>
<p> </p>
</div>
<div id="error-in-proportion-estimate-v.-counts" class="section level2">
<h2><span class="header-section-number">5.4</span> Error in Proportion Estimate v. Counts</h2>
<p>Notice how proportion on its own is a very bad predictor of the uncertainty/error!</p>
<pre class="r"><code>title = &quot;Weak error dependency on proportion value&quot;
ggplot(data = bigdt2, 
       mapping = aes(x = TrueProportion,
                     y = abs(TrueProportion - meanCount/L) / TrueProportion,
                     color = LibrarySize)) + 
  geom_point(size=5, alpha=0.5) + 
  scale_y_log10() + 
  scale_x_log10() + 
  annotation_logticks(sides = &quot;lb&quot;) +
  xlab(&quot;Observed Proportion&quot;) + 
  ylab(&quot;Error [Proportion]&quot;) +
  ggtitle(title)</code></pre>
<p><img src="Lab-03-Poisson-shot-noise_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
<pre class="r"><code>title = &quot;Weak error dependency on proportion&quot;
ggplot(data = bigdt, 
       mapping = aes(x = proportion, 
                     y = log2(proportion / TrueProportion), 
                     colour = LibrarySize)) + 
  geom_hline(yintercept = 0, size = 0.25) +
  geom_point(alpha = 0.5) +
  scale_x_log10() +
  ylab(&quot;Log2(Observed / Truth)&quot;) +
  annotation_logticks(sides = &quot;b&quot;) +
  ggtitle(title)</code></pre>
<p><img src="Lab-03-Poisson-shot-noise_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<p>Meanwhile the uncertainty is actually highly structured with respect to counts (it is Poisson after all).</p>
<pre class="r"><code>title = &quot;Error strongly dependent on count&quot;
ggplot(data = bigdt, 
       # mapping = aes(N, StdErr, colour = LibrarySize)) + 
       mapping = aes(N, log2(proportion / TrueProportion), colour = LibrarySize)) + 
  geom_hline(yintercept = 0, size = 0.25) +
  geom_point(alpha = 0.5) +
  ylab(&quot;Log2(Observed / Truth)&quot;) +
  xlab(&quot;Count&quot;) +
  scale_x_log10() +
  annotation_logticks(sides = &quot;b&quot;) +
  ggtitle(title)</code></pre>
<p><img src="Lab-03-Poisson-shot-noise_files/figure-html/unnamed-chunk-11-1.png" width="672" /></p>
<pre class="r"><code>title = &quot;Error strongly dependent on count&quot;
ggplot(data = bigdt2, 
       mapping = aes(x = meanCount,
                     y = abs(TrueProportion - meanCount/L) / TrueProportion,
                     color=LibrarySize)) + 
  geom_point(size=5, alpha=0.5) + 
  scale_y_log10() + 
  scale_x_log10() + 
  annotation_logticks(sides = &quot;lb&quot;) +
  xlab(&quot;Mean Count&quot;) + 
  ylab(&quot;Error [Proportion]&quot;) +
  ggtitle(title)</code></pre>
<p><img src="Lab-03-Poisson-shot-noise_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
<pre class="r"><code>title = &quot;Variance in Proportion Estimate v. Observed Mean Counts&quot;
ggplot(data = bigdt[, .(meanCount = mean(N),
                        VarProp = var(proportion)), 
                    by = .(LibrarySize, Taxa)], 
       mapping = aes(meanCount, VarProp, colour=LibrarySize)) + 
  geom_point(size=5, alpha=0.5) + 
  # geom_abline(intercept = 0, slope = 1) + 
  scale_y_log10() + 
  scale_x_log10() + 
  annotation_logticks(sides = &quot;lb&quot;) +
  ggtitle(title)</code></pre>
<pre><code>## Warning: Removed 2 rows containing missing values (geom_point).</code></pre>
<p><img src="Lab-03-Poisson-shot-noise_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
<hr />
<p> </p>
</div>
</div>
<div id="conclusion" class="section level1">
<h1><span class="header-section-number">6</span> Conclusion</h1>
<p>The uncertainty due to sampling is actually highly structured with respect to counts, following the Poisson distribution. For intuition, real experiments have shown that technical replicates (different sequencing runs of the same prepared sample) are also well described by the Poisson. It is clear from this simulation that information represented by the count of each taxa in each sample is highly predictive of the uncertainty in the corresponding proportion. Converting to proportions early and ignoring the discrete counts is perilous, especially at low values (e.g. rare taxa).</p>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
