<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="SISMID 2016 Module 14" />


<title>Lab 03: Mixture Models for Testing Differential Abundance</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<link rel="stylesheet"
      href="site_libs/highlight/default.css"
      type="text/css" />
<script src="site_libs/highlight/highlight.js"></script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>

<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.9em;
  padding-left: 5px;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Lab 03</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="http://sismid16m14.bot-tak.org/">
    <span class="fa fa-home"></span>
     
    home
  </a>
</li>
<li>
  <a href="Lab-03-mixture-models-Questions.html">
    <span class="fa fa-question"></span>
     
    Questions
  </a>
</li>
<li>
  <a href="Lab-03-mixture-models-Answers.html">
    <span class="fa fa-exclamation"></span>
     
    Answers
  </a>
</li>
<li>
  <a href="Lab-03-mixture-models-Answers.Rmd">
    <span class="fa fa-gear"></span>
     
    Source
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-newspaper-o"></span>
     
    Further Reading
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Lab-03-Poisson-shot-noise.html">Poisson Intuition Example</a>
    </li>
    <li>
      <a href="http://dx.doi.org/10.1371/journal.pcbi.1003531">Waste Not, Want Not</a>
    </li>
    <li class="dropdown-header">DESeq2</li>
    <li>
      <a href="http://www.biomedcentral.com/content/pdf/gb-2010-11-10-r106.pdf">DESeq2 article</a>
    </li>
    <li>
      <a href="http://bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2 package</a>
    </li>
    <li>
      <a href="http://www.biomedcentral.com/content/pdf/gb-2010-11-10-r106.pdf">DESeq original</a>
    </li>
    <li class="dropdown-header">edgeR</li>
    <li>
      <a href="http://bioinformatics.oxfordjournals.org/content/26/1/139.short">edgeR article</a>
    </li>
    <li>
      <a href="http://nar.oxfordjournals.org/content/early/2014/04/20/nar.gku310.full">edgeR-Robust</a>
    </li>
    <li>
      <a href="http://bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR package</a>
    </li>
    <li class="dropdown-header">metagenomeSeq</li>
    <li>
      <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4010126/">metagenomeSeq article</a>
    </li>
    <li>
      <a href="http://bioconductor.org/packages/release/bioc/html/metagenomeSeq.html">metagenomeSeq package</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Lab 03: Mixture Models for Testing Differential Abundance</h1>
<h4 class="author"><em>SISMID 2016 Module 14</em></h4>
<h4 class="date"><em>Mon Jul 25 14:45:13 2016</em></h4>

</div>


<p> </p>
<div id="topics" class="section level1">
<h1><span class="header-section-number">1</span> Topics</h1>
<style type="text/css">
p { text-align: left; }
</style>
<p>Expected Time to Completion:</p>
<p><strong>30 - 45 minutes</strong></p>
<ul>
<li>DESeq2</li>
<li>edgeR / edgeR-robust</li>
<li>metagenomeSeq</li>
</ul>
<p> </p>
<hr />
</div>
<div id="questions" class="section level1">
<h1><span class="header-section-number">2</span> Questions</h1>
<ul>
<li>Use <code>closedps</code> in the <a href="example-data.RData">included example dataset</a>.</li>
<li><strong>Differential Abundance</strong> - Test for taxa/OTUs that are differentially abundant between the <code>Fast</code> and the <code>Control</code> sample classes. Use <code>phyloseq::mt</code>, <code>DESeq2</code>, <code>edgeR</code>, <code>edgeR-robust</code>, and <code>metagenomeSeq</code>.</li>
<li><strong>Plot</strong> - Plot your results from the differential abundance testing in an informative way that illustrates the confidence of the test result as well as the taxonomic classification. This plot should be more easily interpretable than a table of values.</li>
<li><strong>Compare</strong> - Compare the results from each method. Which taxa/OTUs were consistently rejected (significant)? Which methods tended to reject more taxa/OTUs? Which method would you prefer to use on your own data?</li>
<li><strong>Bonus</strong> - Combine the results of each test method into a single plot.</li>
<li><strong>Variance Stabilization</strong> - Time-allowing, explore ordination results (<code>plot_ordination</code>) after transforming the count data using DESeq2’s <code>getVarianceStabilizedData</code> and <code>rlog</code> functions.</li>
</ul>
<p> </p>
<hr />
</div>
<div id="hints" class="section level1">
<h1><span class="header-section-number">3</span> Hints</h1>
<ul>
<li>There are lots of package-specific idiosyncracies. I have provided example functions to help you quickly convert phyloseq data into the relevant data format. In addition, phyloseq includes a conversion function for DESeq2.</li>
<li>Try to remember to <em>unload</em> one package before you switch to the next, to avoid namespace conflicts. e.g. <code>unloadNamespace(&quot;metagenomeSeq&quot;)</code></li>
<li>Don’t get stuck on plotting. I’m most interested that you are able to successfully run and compare these methods. Organize your results in a <code>data.frame</code>, and then find a <code>ggplot2</code> plot example online.</li>
<li>In case I haven’t mentioned you can replace the <code>otu_table</code> in a phyloseq object using <code>otu_table(physeq) &lt;- x</code>, where <code>physeq</code> is a copy of the phyloseq object you want to modify and <code>x</code> is the new <code>otu_table</code>. This will be very helpful for ordinations on the <code>transformed</code> count values in the <em>variance stabilization</em> question.</li>
</ul>
<p> </p>
<hr />
</div>
<div id="bonus" class="section level1">
<h1><span class="header-section-number">4</span> Bonus</h1>
<ul>
<li>Try an alternative method that isn’t provided (yet) in this lab / tutorial.</li>
</ul>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
